setwd("C:/Users/camnu/bin/salmon-genomic-vulnerability/scripts/tas_inspect_for_ian/")

# library("tidyverse")
source("encode_snps_for_ml.r")

#cd /mnt/c/Users/camnu/bin/salmon-genomic-vulnerability/data/raw/tasmanian_for_ian
tas_ped = "TAS_Salmon_AllFinal.ped"
tas_map = "TAS_Salmon_AllFinal.map"

setwd("C:/Users/camnu/bin/salmon-genomic-vulnerability/data/raw/tasmanian_for_ian/")

tas_data = readPedMap_tsv_fmt(tas_ped, tas_map)

dpath = "C:/Users/camnu/bin/salmon-euro-introgression/data/May2022_SNPchip_220K/SNPchip_100MS_SNPpanel_King7_Compare/SNPchip_220K/"
ped_file_513 = paste0(dpath,"panel_snp_513_set.ped")
map_file_513 = paste0(dpath, "panel_snp_513_set.map")

salmon_513_data = readPedMap_tsv_fmt(ped_file_513, map_file_513)

sum(names(salmon_513_data) %in% names(tas_data))
length(names(salmon_513_data) %in% names(tas_data))

print("only 188 of the markers from the panel present, need to go for conventional pairing and admixture")



#plink --bfile all_fish_Salmo220K_Polyhigh_maf01 --chr-set 30 no-xy --recode --tab --out all_fish_Salmo220K_Polyhigh_maf01
#all_fish_Salmo220K_Polyhigh_maf01



full_dpath = "C:/Users/camnu/bin/salmon-euro-introgression/data/May2022_SNPchip_220K/SNPchip_100MS_SNPpanel_King7_Compare/SNPchip_220K/"
map_file = paste0(full_dpath, "all_fish_Salmo220K_Polyhigh_maf01.map")

map_data = read_tsv(map_file, skip = 1, col_names = c("chromosome", "snp" , "genetic_distance", "physical_distance"))

filter_outfile

rand_sub = sample(names(tas_data), 15000)

overlap = map_data$snp[ map_data$snp %in% rand_sub]
exclude = map_data$snp[!map_data$snp %in% rand_sub]

filter_outfile = "TAS_exclude_list.tsv"
write.table(exclude, filter_outfile, col.names = F, row.names = F, quote = F, sep = "\t")


# cd /Users/camnu/bin/salmon-euro-introgression/data/May2022_SNPchip_220K/SNPchip_100MS_SNPpanel_King7_Compare/SNPchip_220K/
#plink --bfile all_fish_Salmo220K_Polyhigh_maf01 --exclude TAS_exclude_list.tsv --out all_NA_EU_overlap_with_TAS_snps --recode --tab --chr-set 30 no-xy
# move to the working dir


setwd("C:/Users/camnu/bin/salmon-genomic-vulnerability/data/raw/tasmanian_for_ian/")
ped_file = "all_NA_EU_overlap_with_TAS_snps.ped"
map_file = "all_NA_EU_overlap_with_TAS_snps.map"

#~13k SNP sample set chosen from overlap list in taz
salmon_full_data = readPedMap_tsv_fmt(ped_file, map_file)

sum(names(salmon_full_data) %in% names(tas_data))

tas_data = tas_data[names(tas_data) %in% names(salmon_full_data)]
salmon_full_data = salmon_full_data[names(salmon_full_data) %in% names(tas_data)]

major_minor_data = read_table("all_fish_Salmo220K_Polyhigh_maf01.bim", 
                              col_names = c("chr", "snp", "cm",  "pos", "major", "minor"))


tas_data
#bim file has the major and minor from the original, see if A == major and B == minor or if the order has been scrambled

snp_cols = names(tas_data)[7:length(names(tas_data))]


major_match = 0
major_diff = 0
major_org_max = 0
#snp = "AX-87810738"
#snp = "AX-87195613"
for(snp in snp_cols){
  snp_dat = major_minor_data[major_minor_data$snp == snp,]
  
  major = snp_dat[["minor"]]
  maj_gt = paste0(major, " ", major)
  
  
  org_snp = salmon_full_data[snp]
  taz_snp = tas_data[[snp]]

  org_table = table(org_snp)    
  taz_table = table(taz_snp)
  
  org_max = tryCatch({max(org_table) == org_table[[maj_gt]]}, error=function(msg){return(FALSE)})
  taz_max = tryCatch({max(taz_table) == taz_table[["A A"]]}, error=function(msg){return(FALSE)})
  
  if(org_max == TRUE){
    major_org_max = major_org_max+1
    if(taz_max == TRUE){
      major_match = major_match+1
    }
    if(taz_max==FALSE){
      major_diff = major_diff+1
    }
  }
  
}


print(paste0("original major was most seen in:", major_org_max))
print(paste0("matched to AA in: ", major_match ))
print(paste0("did not match to AA in: ",  major_diff))