
#####################################
# Encode SNP data for machine learning 
# - these functions from CamBioR
#####################################
library(tidyverse)

readPedMap_tsv_fmt = function(ped_file, map_file = "", headers = FALSE){
  if(headers == TRUE){
    #
    map_data = read_tsv(map_file, skip = 1, col_names = c("chromosome", "snp" , "genetic_distance", "physical_distance"))
    
    header_data = c('#family', 'individual', 'sire', 'dam', 'sex', 'pheno')
    header_data = c(header_data, map_data$snp)
    
    snp_data = read_tsv(ped_file, skip = 1, col_names = header_data)
    return(snp_data)
    
  }else{
    
    map_data = read_tsv(map_file, col_names = c("chromosome", "snp" , "genetic_distance", "physical_distance"))
    
    header_data = c('#family', 'individual', 'sire', 'dam', 'sex', 'pheno')
    header_data = c(header_data, map_data$snp)
    
    snp_data = read_tsv(ped_file, col_names = header_data)
    return(snp_data)
  }
  
}

get_unique_alleles = function(snp_table){
  unique_alleles = list()

  for(x in names(snp_table)){

    for (a in strsplit(x, " ")[[1]]){
      if( a %in% names(unique_alleles)){
        unique_alleles[[a]] =  unique_alleles[[a]] + snp_table[[x]]
      }else{
        unique_alleles[[a]] =  snp_table[[x]]
      }
    }

  }
  if(length(unique_alleles) > 2){
    stop("SNP is not biallelic")
  }
  if(length(unique_alleles) == 1){
    stop("SNP is homozygous")
  }
  if(unique_alleles[[1]] >= unique_alleles[[2]]){
    return(list("p" = names(unique_alleles)[[1]], "q" = names(unique_alleles)[[2]]))
  }else{
    return(list("p" = names(unique_alleles)[[2]], "q" = names(unique_alleles)[[1]]))
  }

}

#' Homozygous for major allele encoded as 0, heterozygoous = 1, Homozygous minor allele = 2
#' method - make it so you can do one hot, dosage, or presence/absence. currently just dosage
#' missing_data - can put in the mode or NA
encode_dosage_snp_vec = function(snp_vec, method = "dosage", missing_val = "0 0", replace_missing_method = "mode"){
  #determine the major and minor alleles

  if(replace_missing_method == "mode"){

    #replace the missing
    counts = table(snp_vec)

    mode = names(counts[length(counts)])
    if(mode == missing_val){
      warning("most common allele is a missing genotype!")
      mode = names(counts[length(counts)-1])
    }
    snp_vec = replace(snp_vec, snp_vec == missing_val, mode)

  }

  snp_counts = table(snp_vec)
  print(snp_counts)
  alleles = get_unique_alleles(snp_counts)
  p = alleles$p
  q = alleles$q

  AA = paste(p, p)
  AB = paste(p, q)
  BA = paste(q, p) #do this in either order to account for when the major allele here != the major allele from plink
  BB = paste(q, q)

  encodings = c(0, 1, 1, 2)
  names(encodings) = c(AA, AB, BA, BB)

  return(unlist(lapply(snp_vec, function(x){encodings[[x]] })))

}



#'
#' method - make it so you can do one hot, dosage, or presence/absence. currently just dosage
encode_ped = function(snp_data, snp_columns, method = "dosage" ){
  #x = snp_columns[[2]]
  for(x in snp_columns){
    print(x)
    snp_data[[x]] = encode_dosage_snp_vec(snp_data[[x]])
  }

  return(snp_data)
}



###################
# run the analysis
library(tidyverse)
library(data.table)

setwd("C:/Users/camnu/bin/salmon-euro-introgression/data/May2022_SNPchip_220K/SNPchip_100MS_SNPpanel_King7_Compare/SNPchip_220K/")


meta_data = 'panel_data_admix_vals.tsv'
ped_file = 'panel_snp_513_set.ped'
map_file = 'panel_snp_513_set.map'

meta_df = read_tsv(meta_data)

snp_data = readPedMap_tsv_fmt(ped_file, map_file)
snp_columns = names(snp_data)[7:length(names(snp_data))]

length(snp_columns) == 513

snp_data = encode_ped(snp_data, snp_columns)

meta_df$individual = meta_df$fish_id

all_shared = inner_join(meta_df, snp_data)

write_tsv(all_shared, "encoded_snp_panel_with_metadata.tsv")

print("TODO - retain the mode for each column in a second tsv file.")
print("would need this information for future reuse of the model.")
print("it will allow for the imputation to be standardized")

print("also save the ratios currently being printed by the function.")