#########################
# libraries

library(tidyverse)
#########################
# functions

get_wss = function(input_df, PC_colnames = c("PC1", "PC2"), max_k = 10){
  
  pc_only = input_df[PC_colnames]
  
  within_sum_squares = c()
  print("determine optimal k")
  for (i in 1:max_k){
    within_sum_squares[i] = sum(kmeans(pc_only,
                                       centers=i, nstart=25, iter.max=1000)$withinss)
  }
  
  return(within_sum_squares)
}


assign_kmeans = function(input_df, PC_colnames = c("PC1", "PC2"), k = 3){
  
  pc_only = input_df[PC_colnames]
  
  optimal_k = kmeans(pc_only, k, nstart=25, iter.max=1000)
  
  input_df$cluster = as.factor(optimal_k$cluster)
  
  return(input_df)
}



#########################
# load the data
data_path = "C:/Users/camnu/bin/salmon-genomic-vulnerability/data/"
setwd(data_path)


#5455 individuals, all including those not matched to the environment data
full_data_PCA = read_tsv("full_wild_PCA_and_meta.tsv")
full_data_PCA$IID = full_data_PCA$fish_id


#reduced individuals, only those with good matches to the metadata
# matched_data_PCA = read_tsv("interim/220k_wild_meta_match_genetic_PCs_with_ids.tsv")
# matched_metadata = read_tsv("interim/salmo_220K_wildNA_ENVmatchyear_full_Metadata.tsv")
# matched_metadata$IID = matched_metadata$ID
# matched_data_PCA_with_meta = left_join(matched_data_PCA, matched_metadata )
# matched_data_PCA_with_meta_and_prov = left_join(matched_data_PCA_with_meta, full_data_PCA[c("IID", "Province")])
# write_tsv(matched_data_PCA_with_meta_and_prov, "interim/220k_wild_meta_match_genetic_PCs_with_ids_and_metadata.tsv")

matched_data_PCA_with_meta_and_prov = read_tsv("interim/220k_wild_meta_match_genetic_PCs_with_ids_and_metadata.tsv")



#########################
# get wss and plot

print("get wss for k of 1 - 10")
full_wss = get_wss(full_data_PCA)

matched_wss = get_wss(matched_data_PCA_with_meta_and_prov)


print("plot the skree info")

print("full data set")
plot(1:10, full_wss, type="b", xlab="Number of Clusters",
     ylab="Within groups sum of squares", main = "Determining optimal k-means clustering of full individual PCA data")



print("matched data set")
plot(1:10, matched_wss, type="b", xlab="Number of Clusters",
     ylab="Within groups sum of squares", main = "Determining optimal k-means clustering of matched individual PCA data")

print("three appears optimal for both")


##########################
# assign individuals to clusters via k means clustering


clustered_full_data_PCA = assign_kmeans(full_data_PCA)

clustered_matched_data_PCA = assign_kmeans(matched_data_PCA_with_meta_and_prov)


###########################
# rename clusters, save the clusters to file





###########################
# plot the clusters

pcs_clusters = ggplot() + geom_point(data = pcs_plot, aes(x = PC1, y = PC2, colour=cluster))+
  geom_mark_ellipse(data=pcs_plot, aes(x = PC1, y = PC2, fill = cluster,
                                       label = cluster),
                    expand = unit(0.5,"mm"))+
  labs(title =  "Cunner samples, k-means clustering of PCs, 4 groups is optimal", 
       subtitle = "850 individuals, PCA based on full set of SNPs")+
  xlab("PC1, 5.70% variance explained") +
  ylab("PC2, 0.56% variance explained")

ggsave('../external/cunner_optimal_kmeans_PCA.png', pcs_clusters)

write_tsv(pcs_plot, "../external/per_sample_PC_k_cluster_info.tsv")





##############################
##############################
##############################
## starting point code below
##############################
##############################
##############################



setwd("C:/Users/camnu/bin/analysis-cunner-snps/data/raw/")

library(tidyverse)
#install.packages("RcppCNPy")
library(RcppCNPy)
#install.packages("ggforce")
library(ggforce)

library(ggplot2)
theme_set(theme_bw())
library(sf)
library(rnaturalearth)
library(rnaturalearthdata)

library(data.table)
################################################################################

sample_names = data.frame(sample_id=readLines('../../scripts/data/sample_list.txt'))

sample_names$location = unlist(lapply(sample_names$sample_id, function(x){
                                    substr(x, 1,3)}))


datadir = "C:/Users/camnu/bin/analysis-cunner-snps/data/raw/"
outputs = list.files(datadir)


# look at the manual for help with output intrepretation
# https://github.com/Rosemeis/pcangsd
# http://www.popgen.dk/software/index.php/PCAngsd


C = as.matrix(read.table("PC12_merged.cov")) # Reads in estimated covariance matrix
D = npyLoad("PC12_merged.selection.npy") # Reads PC based selection statistics

# Plot PCA plot
e = eigen(C)

#var explained per axis
eigenvalues <- e$values
(eigenvalues[1]/sum(eigenvalues))*100  #5.70% variance explained!? seems real high
(eigenvalues[2]/sum(eigenvalues))*100 #0.563% #using the first two seems good
(eigenvalues[3]/sum(eigenvalues))*100 #0.211%
(eigenvalues[4]/sum(eigenvalues))*100 #0.203%
(eigenvalues[5]/sum(eigenvalues))*100 #0.196%
(eigenvalues[6]/sum(eigenvalues))*100 #0.196%


pcs_plot = data.frame("PC1" = e$vectors[,1], "PC2"= e$vectors[,2])

pcs_plot = cbind(pcs_plot, sample_names)

#plot(e$vectors[,1:2], xlab="PC1", ylab="PC2", main="PCAngsd")
table(pcs_plot$location)

location_pcs_full_all_samples = ggplot() + geom_point(data = pcs_plot, aes(x = PC1, y = PC2, colour=location))+
                        labs(title =  "Cunner samples, PC inspection by location", 
                             subtitle = "850 individuals, PCA based on full set of SNPs")+
                        xlab("PC1, 5.70% variance explained") +
                        ylab("PC2, 0.56% variance explained")

ggsave('../external/cunner_fullPCA_by_location.png', location_pcs_full_all_samples)



pcs_plot = pcs_plot[pcs_plot$location != "DIB", ]
nrow(pcs_plot) #804 total



location_pcs_full_all_samples = ggplot() + geom_point(data = pcs_plot, aes(x = PC1, y = PC2, colour=location))+
  labs(colour =  "Sampling Location")+
  xlab("PC1, 5.70% variance explained") +
  ylab("PC2, 0.56% variance explained")

ggsave('../external/final_cunner_fullPCA_by_location.png', location_pcs_full_all_samples)




################################################################################
# k means clustering of PCA results
comp = data.frame(e$vectors[,1:4])

plot(comp) #pretty linear by PC 2 vs PC 3

pc_only = data.frame("PC1" = e$vectors[,1], "PC2"= e$vectors[,2])

within_sum_squares = c()
print("determine optimal k")
for (i in 1:20){
  within_sum_squares[i] = sum(kmeans(pc_only,
                                     centers=i, nstart=25, iter.max=1000)$withinss)
}

print("plot the skree info")

plot(1:20, within_sum_squares, type="b", xlab="Number of Clusters",
     ylab="Within groups sum of squares", main = "Determining optimal k-means clustering of PCA data")

#looks to be 4 based on the skree plot (inflection point)
optimal_k = kmeans(pc_only, 4, nstart=25, iter.max=1000)

pcs_plot$cluster = as.factor(optimal_k$cluster)

pcs_clusters = ggplot() + geom_point(data = pcs_plot, aes(x = PC1, y = PC2, colour=cluster))+
  geom_mark_ellipse(data=pcs_plot, aes(x = PC1, y = PC2, fill = cluster,
                        label = cluster),
                    expand = unit(0.5,"mm"))+
  labs(title =  "Cunner samples, k-means clustering of PCs, 4 groups is optimal", 
       subtitle = "850 individuals, PCA based on full set of SNPs")+
  xlab("PC1, 5.70% variance explained") +
  ylab("PC2, 0.56% variance explained")

ggsave('../external/cunner_optimal_kmeans_PCA.png', pcs_clusters)

write_tsv(pcs_plot, "../external/per_sample_PC_k_cluster_info.tsv")

#######################################################################################

#looks like 5 distinct clusters, with a single strong outlier populatiion 
#(TUK or STE? need to change colours to tell for sure)

##TODO
# - get the location metadata figure out which are the geographic outliers, geographically similar
# - colour the plot slightly differently (use a is_TUK and is_STE column to get the outliers)
# - read in the metadata and join with the PCs, colour the points by latitude and see if there are trends

# double check with Tony that you pulled the right samples off the final plate. The outliers are strong enough I
# wonder if I have in fact just accidentally thrown some salmon into the genotyping data!


pcs_plot$is_tuk = unlist(lapply(pcs_plot$location, function(x){x == "TUK"}))

TUK_dowsample_PCs_check = ggplot() + geom_point(data = pcs_plot, aes(x = PC1, y = PC2, colour=is_tuk))+
  labs(title =  "Cunner samples, PC inspection by location", 
       subtitle = "Confirming the outlier group is the 53 Tukerton NJ samples")+
  xlab("PC1, 5.70% variance explained") +
  ylab("PC2, 0.56% variance explained")

ggsave('../external/cunner_fullPCA_by_location_Tukerton_outliers_confirm.png', TUK_dowsample_PCs_check)


metadat=read_csv("C:/Users/camnu/bin/cunner-genome/data/raw/Cunner2019_MasterMetadata.csv")
metadat$location = metadat$Code

pcs_plot = left_join(pcs_plot, metadat[c('location', 'Lat', 'Lon')])
pcs_plot


latlon_PCs_check = ggplot() + geom_point(data = pcs_plot, aes(x = PC1, y = PC2, colour=Lat))+
  labs(title =  "Cunner samples, PC with points coloured by latitude", 
       subtitle = "suggestive of some clustering by latitude/geographic proximity")+
  xlab("PC1, 5.70% variance explained") +
  ylab("PC2, 0.56% variance explained")

ggsave('../external/cunner_fullPCA_by_latitude.png', latlon_PCs_check)


non_nj = pcs_plot[pcs_plot$is_tuk == FALSE, ]

nonnj_latlon_PCs_check = ggplot() + geom_point(data = non_nj, aes(x = PC1, y = PC2, colour=Lat))+
  labs(title =  "Cunner samples, PC inspection by location, latitude", 
       subtitle = "removing the NJ samples to zoom in")+
  xlab("PC1") +
  ylab("PC2")

ggsave('../external/cunner_fullPCA_by_latitude_NJ_rm.png', nonnj_latlon_PCs_check)


nonnj_location_PCs_check = ggplot() + geom_point(data = non_nj, aes(x = PC1, y = PC2, colour=location))+
  labs(title =  "Cunner samples, PC inspection by location", 
       subtitle = "removing the NJ samples to zoom in")+
  xlab("PC1") +
  ylab("PC2")

ggsave('../external/cunner_fullPCA_by_location_NJ_rm.png', nonnj_location_PCs_check)



non_nj$is_dib =non_nj$location == "DIB"

digby_latlon_PCs_check = ggplot() + geom_point(data = non_nj, aes(x = PC1, y = PC2, colour=is_dib))+
  labs(title =  "Cunner PC inspection, Digby samples isolated", 
       subtitle = "Digby were F1 of wild crosses, so possibly some relatedness causing this.")+
  xlab("PC1") +
  ylab("PC2")

ggsave('../external/cunner_fullPCA_by_latitude_NJ_rm_digby_iso.png', digby_latlon_PCs_check)


#######################################################################################
# plot the PC clusters on map to see if there are any patterns of note


cluster_dat = read_tsv("../external/per_sample_PC_k_cluster_info.tsv")

metadat=read_csv("C:/Users/camnu/bin/cunner-genome/data/raw/Cunner2019_MasterMetadata.csv")
metadat$location = metadat$Code

pcs_plot = left_join(cluster_dat, metadat[c('location', 'Lat', 'Lon')])


#BHN samples - lack metadata?


###################################
# search for evidence of genetic / geographic location interactions.


world = ne_countries(scale = "medium", returnclass = "sf")
class(world)
#plot a sub region
lat_min = min(metadat$Lat)
lat_max = max(metadat$Lat)
long_min = min(metadat$Lon)
long_max = max(metadat$Lon)

clusters_by_study_range = ggplot(data = world) +
  geom_sf() +
  coord_sf(xlim = c(long_min - 2, long_max + 2), 
           ylim = c(lat_min - 2, lat_max + 2), expand = FALSE) +
  labs(title = "Cunner individuals, coloured by their k-means cluster of PCs", 
       subtitle = "points jittered to prevent overlap")+
  geom_jitter(data = pcs_plot, aes(x = Lon, y = Lat, 
                                   colour=as.factor(cluster)), 
              position=position_jitter(width = .25, height = 0.25), alpha = 0.5)

  
ggsave('../../data/external/cunner_sample_cluster_locations.png', clusters_by_study_range)


# table(pcs_plot$cluster, pcs_plot$location)


################################################################################
# additional checks, data log and metadata

metadat=read_csv("C:/Users/camnu/bin/cunner-genome/data/raw/Cunner2019_MasterMetadata.csv")

processing_log = read_tsv("C:/Users/camnu/bin/analysis-cunner-snps/data/cunner_data_log.tsv")


# 0. BHN samples - lack metadata?
# resolved, were missing a field in one of the excel entries BHN == bar haven north 
pcs_plot[pcs_plot$location == "BHN",] #they're added to the map now



# 1 . assume these individuals were randomized across plates / flow cells and we are not seeing some sort of processing artefact here.
# No, see below
names(processing_log)

dib_seq_dat = processing_log[processing_log$location == "DIB",]

dib_seq_dat$R1_source

table(dib_seq_dat$batch_id) 
#some DIB in each of my 9 different processing batches, so not a binf processing effect as they were handled separately.

pcs_plot[pcs_plot$location == "DIB",]

# unfortunately it looks like all 46 digby samples were processed on 2021-10-06 as part of sequencing run 1675
# so can't compare plates or sequencing runs to find if one of them is the source of the problem, and therefore
# can't rule out a lab artefact.
# as you stated, for next time its probably best if the samples are randomized across plates.



#2 . Are the sites in Placentia Bay all south coast? 
#Hard to tell from the map. If so then the NL groups are making some sense and in some ways similar to other species and the local environment.  

#In placentia bay:			
#Bar Haven North, NL	47.70929103	-54.21512696
#Arnolds Cove, Placentia, NL	47.758691	-53.990043

# Yes, both fall south on the south coast



#3. looking genome wide to see if this association is wide spread or being driven by a a specific region

# see the analyze_fst.r outputs
# between certain populations (1v3, 1v4) there are small differences
# between 3 and 4 differences on every chromosome it seems
# between 1 and 2 the whole genome seems different


#4. As a followup for looking for peaks of differentiation, maybe try plotting the loadings from the PCAngsd analysis 
#with all of the SNPs run - there's code in the Salmon sea age repo already.
#Might give some clues as to whether there are clear signals of adaptation across the genome. 
#The presence of clear peaks (similar to a GWAS) would suggest the pop structure observed here is due to some degree of local adaptation
#also a chance there is something obvious and cool differentiating these clusters.



# to do the loadings, look at this line:
#https://github.com/TonyKess/seaage_GWAS/blob/dd2e1b8d38d28a02b2289e76ec1675bd77eab574/analysis.1.FST_GWAS_rdachipcomp.R#L153
# tony reads in the loadings file and then runs some analyses on it.

D_loadings = npyLoad("PC12_merged.selection.npy") # Reads PC based selection statistics

all_sites = fread("PC12_merged.sites", data.table = F,
                   stringsAsFactors = F, header = F)

outputs
