
############
# load libraries

library(tidyverse)


############
# set path
data_path = "C:/Users/camnu/bin/salmon-genomic-vulnerability/data/"
setwd(data_path)


############
# functions

readPedMap_tsv_fmt = function(ped_file, map_file = "", headers = FALSE){
  if(headers == TRUE){
    #
    map_data = read_tsv(map_file, skip = 1, col_names = c("chromosome", "snp" , "genetic_distance", "physical_distance"))
    
    header_data = c('#family', 'individual', 'sire', 'dam', 'sex', 'pheno')
    header_data = c(header_data, map_data$snp)
    
    snp_data = read_tsv(ped_file, skip = 1, col_names = header_data)
    return(snp_data)
    
  }else{
    
    map_data = read_tsv(map_file, col_names = c("chromosome", "snp" , "genetic_distance", "physical_distance"))
    
    header_data = c('#family', 'individual', 'sire', 'dam', 'sex', 'pheno')
    header_data = c(header_data, map_data$snp)
    
    snp_data = read_tsv(ped_file, col_names = header_data)
    return(snp_data)
  }
  
}



############
# read the data

clustered_full_data_PCA = read_tsv("interim/full_data_PCA_kmeans_cluster_info.tsv")
clustered_matched_data_PCA = read_tsv("interim/matched_data_PCA_kmeans_cluster_info.tsv")


print("build the phenotype file")

ped_data = readPedMap_tsv_fmt("raw/wildNA_220k_vulnerability_work.ped", "raw/wildNA_220k_vulnerability_work.map")
ped_data = ped_data[c('#family', 'individual', 'sire', 'dam', 'sex', 'pheno')]
gc()


print("make two pheno files from the above, one for the reduced data set (with clusters based off that) and one for the full")

clustered_full_data_PCA$individual = clustered_full_data_PCA$fish_id
clustered_matched_data_PCA$individual = clustered_matched_data_PCA$IID

pheno_full = left_join(ped_data, clustered_full_data_PCA[c("individual", "cluster", "cluster_name")])
pheno_matched = left_join(ped_data, clustered_matched_data_PCA[c("individual", "cluster", "cluster_name")])
pheno_matched = na.omit(pheno_matched) #drop the non matching rows

code_cluster = c("LAB", "NFL", "STL")
names(code_cluster) = c("Labrador", "Newfoundland", "St. Lawrence")

pheno_full$cluster_code = unlist(lapply(pheno_full$cluster_name, function(x)code_cluster[[x]]))
table(pheno_full$cluster_name)
table(pheno_full$cluster_code)

pheno_matched$cluster_code = unlist(lapply(pheno_matched$cluster_name, function(x)code_cluster[[x]]))


pheno_full = pheno_full[c('#family', 'individual', "cluster_code")]
names(pheno_full) = c("FID", "IID", "CID")
pheno_matched = pheno_matched[c('#family', 'individual', "cluster_code")]
names(pheno_matched) = c("FID", "IID", "CID")


write_tsv(pheno_full, "raw/")

write_tsv(pheno_matched, "raw/")


############
# build the plink file with the pheno info 

#https://www.cog-genomics.org/plink/2.0/basic_stats
# see pairwise fixation index
