
############
# load libraries

library(tidyverse)


############
# set path
data_path = "C:/Users/camnu/bin/salmon-genomic-vulnerability/data/"
setwd(data_path)


############
# functions

readPedMap_tsv_fmt = function(ped_file, map_file = "", headers = FALSE){
  if(headers == TRUE){
    #
    map_data = read_tsv(map_file, skip = 1, col_names = c("chromosome", "snp" , "genetic_distance", "physical_distance"))
    
    header_data = c('#family', 'individual', 'sire', 'dam', 'sex', 'pheno')
    header_data = c(header_data, map_data$snp)
    
    snp_data = read_tsv(ped_file, skip = 1, col_names = header_data)
    return(snp_data)
    
  }else{
    
    map_data = read_tsv(map_file, col_names = c("chromosome", "snp" , "genetic_distance", "physical_distance"))
    
    header_data = c('#family', 'individual', 'sire', 'dam', 'sex', 'pheno')
    header_data = c(header_data, map_data$snp)
    
    snp_data = read_tsv(ped_file, col_names = header_data)
    return(snp_data)
  }
  
}



############
# read the data

clustered_full_data_PCA = read_tsv("interim/full_data_PCA_kmeans_cluster_info.tsv")
clustered_matched_data_PCA = read_tsv("interim/matched_data_PCA_kmeans_cluster_info.tsv")


print("build the phenotype file")

ped_data = readPedMap_tsv_fmt("raw/wildNA_220k_vulnerability_work.ped", "raw/wildNA_220k_vulnerability_work.map")
ped_data = ped_data[c('#family', 'individual', 'sire', 'dam', 'sex', 'pheno')]
gc()


print("make two pheno files from the above, one for the reduced data set (with clusters based off that) and one for the full")

clustered_full_data_PCA$individual = clustered_full_data_PCA$fish_id
clustered_matched_data_PCA$individual = clustered_matched_data_PCA$IID

pheno_full = left_join(ped_data, clustered_full_data_PCA[c("individual", "cluster", "cluster_name")])
pheno_matched = left_join(ped_data, clustered_matched_data_PCA[c("individual", "cluster", "cluster_name")])
pheno_matched = na.omit(pheno_matched) #drop the non matching rows

code_cluster = c("LAB", "NFL", "STL")
names(code_cluster) = c("Labrador", "Newfoundland", "St. Lawrence")

pheno_full$cluster_code = unlist(lapply(pheno_full$cluster_name, function(x)code_cluster[[x]]))
table(pheno_full$cluster_name)
table(pheno_full$cluster_code)

pheno_matched$cluster_code = unlist(lapply(pheno_matched$cluster_name, function(x)code_cluster[[x]]))


pheno_full = pheno_full[c('#family', 'individual', "cluster_code")]
names(pheno_full) = c("FID", "IID", "CID")
pheno_matched = pheno_matched[c('#family', 'individual', "cluster_code")]
names(pheno_matched) = c("FID", "IID", "CID")


write_tsv(pheno_full, "raw/full_individual_cluster.pheno")
write_tsv(pheno_matched, "raw/matched_individual_cluster.pheno")

# print("dosage encode the pheno groups, jic thats needed for plink fst")
# 
# for(c in code_cluster){
#   pheno_full[c] = as.integer(unlist(lapply(pheno_full$CID, function(x) x == c)))  
#   pheno_matched[c] = as.integer(unlist(lapply(pheno_matched$CID, function(x) x == c)))  
# }
# 
# pheno_full
# 
# write_tsv(pheno_full[c("FID", "IID", "LAB", "NFL", "STL")], "raw/dosage_full_individual_cluster.pheno")
# write_tsv(pheno_matched[c("FID", "IID", "LAB", "NFL", "STL")], "raw/dosage_matched_individual_cluster.pheno")



############
# make the plink files for the different pairwise comparisons

print("need the different 'keep' files (FMT: FID IID)")


code_cluster = c("LAB", "NFL", "STL")

print("subsettting the different pheno files and prepping for the plink code")
for(i1 in 1:2){
  for(i2 in 2:3){
    if(i1 != i2){
      c1 = code_cluster[i1]
      c2 = code_cluster[i2]
      combo = paste0(c1,"_v_",c2)
      print(paste0("on comparison: ",combo))
      combos = c(c1, c2)
      
      sub_pheno_full = pheno_full[pheno_full$CID %in% combos,]
      write_tsv(sub_pheno_full[c("FID", "IID", "CID")], paste0("raw/fst_files/", combo ,"_full_individual_cluster.pheno"))
      write_tsv(sub_pheno_full[c("FID", "IID")], paste0("raw/fst_files/", combo ,"_full_individual_list.txt"))
      
      sub_pheno_matched = pheno_matched[pheno_matched$CID %in% combos,]
      write_tsv(sub_pheno_matched[c("FID", "IID", "CID")], paste0("raw/fst_files/",combo, "_matched_individual_cluster.pheno"))
      write_tsv(sub_pheno_matched[c("FID", "IID")], paste0("raw/fst_files/",combo, "_matched_individual_list.txt"))
      
    }
  }
}



print("use the files to subset the full bed in plink - make the pairwise compare files")
#plink --bfile wildNA_220k_filtered --chr-set 30 no-xy --keep fst_files/LAB_v_NFL_full_individual_list.txt --make-bed --out fst_files/LAB_v_NFL_full_compare
#plink --bfile wildNA_220k_filtered --chr-set 30 no-xy --keep fst_files/LAB_v_STL_full_individual_list.txt --make-bed --out fst_files/LAB_v_STL_full_compare
#plink --bfile wildNA_220k_filtered --chr-set 30 no-xy --keep fst_files/NFL_v_STL_full_individual_list.txt --make-bed --out fst_files/NFL_v_STL_full_compare

#plink --bfile Salmo_220K_Merged2022_COMPLETE_ENVmatchyear --chr-set 30 no-xy --keep fst_files/LAB_v_NFL_matched_individual_list.txt --make-bed --out fst_files/LAB_v_NFL_matched_compare
#plink --bfile Salmo_220K_Merged2022_COMPLETE_ENVmatchyear --chr-set 30 no-xy --keep fst_files/LAB_v_STL_matched_individual_list.txt --make-bed --out fst_files/LAB_v_STL_matched_compare
#plink --bfile Salmo_220K_Merged2022_COMPLETE_ENVmatchyear --chr-set 30 no-xy --keep fst_files/NFL_v_STL_matched_individual_list.txt --make-bed --out fst_files/NFL_v_STL_matched_compare


print("run the fst for the full set first")
#full data pairwise comparisons
#plink --bfile LAB_v_NFL_full_compare --chr-set 30 no-xy --fst --within LAB_v_NFL_full_individual_cluster.pheno --out LAB_v_NFL_full_pairwise_fst_out
#plink --bfile LAB_v_STL_full_compare --chr-set 30 no-xy --fst --within LAB_v_STL_full_individual_cluster.pheno --out LAB_v_STL_full_pairwise_fst_out
#plink --bfile NFL_v_STL_full_compare --chr-set 30 no-xy --fst --within NFL_v_STL_full_individual_cluster.pheno --out NFL_v_STL_full_pairwise_fst_out

#matched data pw comps
#plink --bfile LAB_v_NFL_matched_compare --chr-set 30 no-xy --fst --within LAB_v_NFL_matched_individual_cluster.pheno --out LAB_v_NFL_matched_pairwise_fst_out
#plink --bfile LAB_v_STL_matched_compare --chr-set 30 no-xy --fst --within LAB_v_STL_matched_individual_cluster.pheno --out LAB_v_STL_matched_pairwise_fst_out
#plink --bfile NFL_v_STL_matched_compare --chr-set 30 no-xy --fst --within NFL_v_STL_matched_individual_cluster.pheno --out NFL_v_STL_matched_pairwise_fst_out


############
# Read in the outputs and assess the differences