
############
# load libraries

library(tidyverse)
library(ggman)


############
# set path
data_path = "C:/Users/camnu/bin/salmon-genomic-vulnerability/data/"
setwd(data_path)


############
# functions

readPedMap_tsv_fmt = function(ped_file, map_file = "", headers = FALSE){
  if(headers == TRUE){
    #
    map_data = read_tsv(map_file, skip = 1, col_names = c("chromosome", "snp" , "genetic_distance", "physical_distance"))
    
    header_data = c('#family', 'individual', 'sire', 'dam', 'sex', 'pheno')
    header_data = c(header_data, map_data$snp)
    
    snp_data = read_tsv(ped_file, skip = 1, col_names = header_data)
    return(snp_data)
    
  }else{
    
    map_data = read_tsv(map_file, col_names = c("chromosome", "snp" , "genetic_distance", "physical_distance"))
    
    header_data = c('#family', 'individual', 'sire', 'dam', 'sex', 'pheno')
    header_data = c(header_data, map_data$snp)
    
    snp_data = read_tsv(ped_file, col_names = header_data)
    return(snp_data)
  }
  
}



############
# read the data

clustered_full_data_PCA = read_tsv("interim/full_data_PCA_kmeans_cluster_info.tsv")
clustered_matched_data_PCA = read_tsv("interim/matched_data_PCA_kmeans_cluster_info.tsv")


print("build the phenotype file")

ped_data = readPedMap_tsv_fmt("raw/wildNA_220k_vulnerability_work.ped", "raw/wildNA_220k_vulnerability_work.map")
ped_data = ped_data[c('#family', 'individual', 'sire', 'dam', 'sex', 'pheno')]
gc()


print("make two pheno files from the above, one for the reduced data set (with clusters based off that) and one for the full")

clustered_full_data_PCA$individual = clustered_full_data_PCA$fish_id
clustered_matched_data_PCA$individual = clustered_matched_data_PCA$IID

pheno_full = left_join(ped_data, clustered_full_data_PCA[c("individual", "cluster", "cluster_name")])
pheno_matched = left_join(ped_data, clustered_matched_data_PCA[c("individual", "cluster", "cluster_name")])
pheno_matched = na.omit(pheno_matched) #drop the non matching rows

code_cluster = c("LAB", "NFL", "STL")
names(code_cluster) = c("Labrador", "Newfoundland", "St. Lawrence")

pheno_full$cluster_code = unlist(lapply(pheno_full$cluster_name, function(x)code_cluster[[x]]))
table(pheno_full$cluster_name)
table(pheno_full$cluster_code)

pheno_matched$cluster_code = unlist(lapply(pheno_matched$cluster_name, function(x)code_cluster[[x]]))


pheno_full = pheno_full[c('#family', 'individual', "cluster_code")]
names(pheno_full) = c("FID", "IID", "CID")
pheno_matched = pheno_matched[c('#family', 'individual', "cluster_code")]
names(pheno_matched) = c("FID", "IID", "CID")


write_tsv(pheno_full, "raw/full_individual_cluster.pheno")
write_tsv(pheno_matched, "raw/matched_individual_cluster.pheno")

# print("dosage encode the pheno groups, jic thats needed for plink fst")
# 
# for(c in code_cluster){
#   pheno_full[c] = as.integer(unlist(lapply(pheno_full$CID, function(x) x == c)))  
#   pheno_matched[c] = as.integer(unlist(lapply(pheno_matched$CID, function(x) x == c)))  
# }
# 
# pheno_full
# 
# write_tsv(pheno_full[c("FID", "IID", "LAB", "NFL", "STL")], "raw/dosage_full_individual_cluster.pheno")
# write_tsv(pheno_matched[c("FID", "IID", "LAB", "NFL", "STL")], "raw/dosage_matched_individual_cluster.pheno")



############
# make the plink files for the different pairwise comparisons

print("need the different 'keep' files (FMT: FID IID)")


code_cluster = c("LAB", "NFL", "STL")

print("subsettting the different pheno files and prepping for the plink code")
for(i1 in 1:2){
  for(i2 in 2:3){
    if(i1 != i2){
      c1 = code_cluster[i1]
      c2 = code_cluster[i2]
      combo = paste0(c1,"_v_",c2)
      print(paste0("on comparison: ",combo))
      combos = c(c1, c2)
      
      sub_pheno_full = pheno_full[pheno_full$CID %in% combos,]
      write_tsv(sub_pheno_full[c("FID", "IID", "CID")], paste0("raw/fst_files/", combo ,"_full_individual_cluster.pheno"))
      write_tsv(sub_pheno_full[c("FID", "IID")], paste0("raw/fst_files/", combo ,"_full_individual_list.txt"))
      
      sub_pheno_matched = pheno_matched[pheno_matched$CID %in% combos,]
      write_tsv(sub_pheno_matched[c("FID", "IID", "CID")], paste0("raw/fst_files/",combo, "_matched_individual_cluster.pheno"))
      write_tsv(sub_pheno_matched[c("FID", "IID")], paste0("raw/fst_files/",combo, "_matched_individual_list.txt"))
      
    }
  }
}



print("use the files to subset the full bed in plink - make the pairwise compare files")
#plink --bfile wildNA_220k_filtered --chr-set 30 no-xy --keep fst_files/LAB_v_NFL_full_individual_list.txt --make-bed --out fst_files/LAB_v_NFL_full_compare
#plink --bfile wildNA_220k_filtered --chr-set 30 no-xy --keep fst_files/LAB_v_STL_full_individual_list.txt --make-bed --out fst_files/LAB_v_STL_full_compare
#plink --bfile wildNA_220k_filtered --chr-set 30 no-xy --keep fst_files/NFL_v_STL_full_individual_list.txt --make-bed --out fst_files/NFL_v_STL_full_compare

#plink --bfile Salmo_220K_Merged2022_COMPLETE_ENVmatchyear --chr-set 30 no-xy --keep fst_files/LAB_v_NFL_matched_individual_list.txt --make-bed --out fst_files/LAB_v_NFL_matched_compare
#plink --bfile Salmo_220K_Merged2022_COMPLETE_ENVmatchyear --chr-set 30 no-xy --keep fst_files/LAB_v_STL_matched_individual_list.txt --make-bed --out fst_files/LAB_v_STL_matched_compare
#plink --bfile Salmo_220K_Merged2022_COMPLETE_ENVmatchyear --chr-set 30 no-xy --keep fst_files/NFL_v_STL_matched_individual_list.txt --make-bed --out fst_files/NFL_v_STL_matched_compare


print("run the fst for the full set first")
#full data pairwise comparisons
#plink --bfile LAB_v_NFL_full_compare --chr-set 30 no-xy --fst --within LAB_v_NFL_full_individual_cluster.pheno --out LAB_v_NFL_full_pairwise_fst_out
#plink --bfile LAB_v_STL_full_compare --chr-set 30 no-xy --fst --within LAB_v_STL_full_individual_cluster.pheno --out LAB_v_STL_full_pairwise_fst_out
#plink --bfile NFL_v_STL_full_compare --chr-set 30 no-xy --fst --within NFL_v_STL_full_individual_cluster.pheno --out NFL_v_STL_full_pairwise_fst_out

#matched data pw comps
#plink --bfile LAB_v_NFL_matched_compare --chr-set 30 no-xy --fst --within LAB_v_NFL_matched_individual_cluster.pheno --out LAB_v_NFL_matched_pairwise_fst_out
#plink --bfile LAB_v_STL_matched_compare --chr-set 30 no-xy --fst --within LAB_v_STL_matched_individual_cluster.pheno --out LAB_v_STL_matched_pairwise_fst_out
#plink --bfile NFL_v_STL_matched_compare --chr-set 30 no-xy --fst --within NFL_v_STL_matched_individual_cluster.pheno --out NFL_v_STL_matched_pairwise_fst_out



############
# Read in the outputs and assess the differences

print("visualize the fst and get the summary stats")
#NB - NMISS == the number of genotype calls considered (bad acronym, implies number missing to me but this is not the case)

comps = c("LAB_v_NFL", "LAB_v_STL", "NFL_v_STL")

full_fst_files = c("raw/fst_files/LAB_v_NFL_full_pairwise_fst_out.fst", 
                      "raw/fst_files/LAB_v_STL_full_pairwise_fst_out.fst", 
                      "raw/fst_files/NFL_v_STL_full_pairwise_fst_out.fst")
names(full_fst_files) = comps

matched_fst_files = c("raw/fst_files/LAB_v_NFL_matched_pairwise_fst_out.fst", 
                      "raw/fst_files/LAB_v_STL_matched_pairwise_fst_out.fst", 
                      "raw/fst_files/NFL_v_STL_matched_pairwise_fst_out.fst")
names(matched_fst_files) = comps


fst_summary = c()
#c = "LAB_v_NFL"
#c = "LAB_v_STL"
#c ="NFL_v_STL"
for(i in 1:length(comps)){
  c = comps[[i]]
  print(paste0("on comp: ", c))
  print("read the data")
  full_fst = read_tsv(full_fst_files[[c]])
  matched_fst = read_tsv(matched_fst_files[[c]])
  
  print("make the manhattan plots")
  full_fst_plot = ggman(full_fst, snp = "SNP", bp = "POS", chrom = "CHR", pvalue = "FST",
                        logTransform = FALSE, ymin = 0, ymax = 1)+
    labs(title = paste0("Full individual set, Fst values comparison: " , c))+
    xlab("Chromosome") +
    ylab("Fst value")+
    geom_hline(yintercept=0.15, linetype="dashed", colour="red", 1)
  
  matched_fst_plot = ggman(matched_fst, snp = "SNP", bp = "POS", chrom = "CHR", pvalue = "FST",
                        logTransform = FALSE, ymin = 0, ymax = 1)+
    labs(title = paste0("Matched individual set, Fst values comparison: " , c))+
    xlab("Chromosome") +
    ylab("Fst value")+
    geom_hline(yintercept=0.15, linetype="dashed", colour="red", 1)

  
  print("write to files")
  ggsave(paste0('external/fst_plots/full_individuals_fst_',c,'.png'), full_fst_plot)
  ggsave(paste0('external/fst_plots/matched_individuals_fst_',c,'.png'), matched_fst_plot)
  
  print("save the summary statistics")
  
  full_count_unique = sum(full_fst$FST == 1,  na.rm = TRUE)
  full_fst_mean = mean(full_fst$FST,  na.rm = TRUE)
  full_fst_min = min(full_fst$FST,  na.rm = TRUE)
  full_fst_max = max(full_fst$FST,  na.rm = TRUE)
  full_count_high = sum(full_fst$FST > .15,  na.rm = TRUE)
  
  print("adding full line to summary data frame")
  fst_summary = rbind(fst_summary ,
                      data.frame(comparison = c,
                                 data_set = "full",
                                 count_unique = full_count_unique,
                                 fst_mean = full_fst_mean,
                                 fst_min = full_fst_min,
                                 fst_max = full_fst_max,
                                 n_fst_greater_15 = full_count_high))
  

  matched_count_unique = sum(matched_fst$FST == 1, na.rm = TRUE)
  matched_fst_mean = mean(matched_fst$FST,  na.rm = TRUE)
  matched_fst_min = min(matched_fst$FST ,  na.rm = TRUE)
  matched_fst_max = max(matched_fst$FST,  na.rm = TRUE)
  matched_count_high = sum(matched_fst$FST > .15,  na.rm = TRUE)
  
  print("adding matched line to summary data frame")
  fst_summary = rbind(fst_summary ,
                      data.frame(comparison = c,
                                 data_set = "matched",
                                 count_unique = matched_count_unique,
                                 fst_mean = matched_fst_mean,
                                 fst_min = matched_fst_min,
                                 fst_max = matched_fst_max,
                                 n_fst_greater_15 = matched_count_high))
}


write_tsv(fst_summary, "external/fst_plots/fst_summary_stats.tsv")





print("extract just the significant locations from the data")

comps = c("LAB_v_NFL", "LAB_v_STL", "NFL_v_STL")

full_fst_files = c("raw/fst_files/LAB_v_NFL_full_pairwise_fst_out.fst", 
                   "raw/fst_files/LAB_v_STL_full_pairwise_fst_out.fst", 
                   "raw/fst_files/NFL_v_STL_full_pairwise_fst_out.fst")
names(full_fst_files) = comps

all_sig_fst = data.frame()

c = "LAB_v_NFL"
for(c in comps){
  input = full_fst_files[[c]]
  full_fst = read_tsv(input)
  
  full_fst$comparison = c
  
  full_fst = full_fst[full_fst$FST > 0.15,]
  
  all_sig_fst = rbind(all_sig_fst, full_fst)
}


write_tsv(all_sig_fst, "interim/all_pw_fst_gte_15.tsv")
