
############
# load libraries

library(tidyverse)
library(ggman)
theme_set(theme_bw())


############
# set path
data_path = "C:/Users/camnu/bin/salmon-genomic-vulnerability/data/"
setwd(data_path)



###########
# get individual lists to subset the plink files


clustered_full_data_PCA = read_tsv("interim/full_data_PCA_kmeans_cluster_info.tsv")
#clustered_matched_data_PCA = read_tsv("interim/matched_data_PCA_kmeans_cluster_info.tsv")

cluster_names = c("Labrador", "Newfoundland", "St. Lawrence")

#c = "Labrador"
for(c in cluster_names){
  
  sub_df = clustered_full_data_PCA[clustered_full_data_PCA$cluster_name == c,]
  
  out_df = sub_df[c("SiteCode", "fish_id")]

  names(out_df) = c("FID", "IID")
  

  write_tsv(out_df, paste0("raw/LD_files/", c, "_individual_list.tsv"))  
}


############
# plink code to subset the bam files and make the smaller individual sets
print("subset the clusters")
#435 samples:
#plink --bfile wildNA_220k_filtered --chr-set 30 no-xy --maf 0.01 --make-bed --keep LD_files/Labrador_individual_list.tsv --out LD_files/Labrador_220k_filtered

#3244 samples:
#plink --bfile wildNA_220k_filtered --chr-set 30 no-xy --maf 0.01 --make-bed --keep LD_files/Newfoundland_individual_list.tsv  --out LD_files/Newfoundland_220k_filtered

#1776 samples:
#plink --bfile wildNA_220k_filtered --chr-set 30 no-xy --maf 0.01 --make-bed --keep LD_files/STLawrence_individual_list.tsv  --out LD_files/STLawrence_220k_filtered

print("run the LD analysis")
##full
###plink --bfile wildNA_220k_filtered --chr-set 30 no-xy --ld-window-kb 1000 --r2 --out LD_files/full_220k_filtered_LD_out
#REDO with updated genome:
# plink --bfile new_noLDfilter_220k_filtered --chr-set 30 no-xy --ld-window-kb 1000 --r2 --out LD_files/new_full_220k_filtered_LD_out

##435 samples:
###plink --bfile LD_files/Labrador_220k_filtered --chr-set 30 no-xy --ld-window-kb 1000 --r2 --out LD_files/Labrador_LD_output
#REDO with updated genome:
# plink --bfile LD_files/new_noLDfilter_Labrador_220k_filtered --chr-set 30 no-xy --ld-window-kb 1000 --r2 --out LD_files/new_Labrador_LD_output

##3244 samples:
###plink --bfile LD_files/Newfoundland_220k_filtered --chr-set 30 no-xy --ld-window-kb 1000 --r2 --out LD_files/Newfoundland_LD_output
#REDO with updated genome:
# plink --bfile LD_files/new_noLDfilter_Newfoundland_220k_filtered --chr-set 30 no-xy --ld-window-kb 1000 --r2 --out LD_files/new_Newfoundland_LD_output

##1776 samples:
###plink --bfile LD_files/STLawrence_220k_filtered --chr-set 30 no-xy --ld-window-kb 1000 --r2 --out LD_files/STLawrence_LD_output
#REDO with updated genome:
# plink --bfile LD_files/new_noLDfilter_STLawrence_220k_filtered --chr-set 30 no-xy --ld-window-kb 1000 --r2 --out LD_files/new_STLawrence_LD_output


##################
# python for summary, already have a script
print("run the windowed summarizer")
#python ld_summary.py


#################
# import and visualize the LD summary info
plot_LD_summary = function(data){
  
  window_plot = ggplot() +
    geom_line(data = data, aes(y = mean_r2, x = win_start), colour = 'grey') +
    geom_point(data = data, aes(y = mean_r2, x = win_start), colour = 'black') +
    labs(title =  paste0("Chromosome ", data$chromosome[[1]], " windowed LD data, ", data$pop[[1]]), 
         subtitle = "")+
    xlab("base pair window start") +
    ylab("Mean r^2 for markers in the 100kb window")
  
  
  outfile = paste0('../external/ld_outputs/LD_summary_plot_', data$pop[[1]], '_chr_', data$chromosome[[1]], '.png')
  ggsave(outfile, window_plot)
  
}


ld_df = read_tsv("raw/LD_files/ld_summary_output.tsv")

table(ld_df$cluster)
table(ld_df$chromosome)

#c = "full"
for(c in unique(ld_df$cluster)){
  sub_df = ld_df[ld_df$cluster == c,]

  
  print("make the manhattan plots")
  median_ld_plot = ggman(sub_df, snp = "win_end", bp = "win_start", chrom = "chromosome", pvalue = "median_r2",
                        logTransform = FALSE, ymin = 0, ymax = 1)+
    labs(title = paste0("Data set: " , c,", 1mb median windowed LD scores"))+
    xlab("Chromosome") +
    ylab("Median R2")
  
  mean_ld_plot = ggman(sub_df, snp = "win_end", bp = "win_start", chrom = "chromosome", pvalue = "mean_r2",
                         logTransform = FALSE, ymin = 0, ymax = 1)+
    labs(title = paste0("Data set: " , c,", 1mb mean windowed LD scores"))+
    xlab("Chromosome") +
    ylab("Median R2")
  
  print("write to files")
  ggsave(paste0('external/ld_plots/cluster_', c,'_median_1mb_LD','.png'), median_ld_plot)
  ggsave(paste0('external/ld_plots/cluster_', c,'_mean_1mb_LD','.png'), mean_ld_plot)
  
  print("make the histograms")
    median_ld_hist = ggplot(sub_df, aes(x=median_r2))+
                      geom_histogram()+
                      labs(title = paste0("Data set: " , c,", genome-wide histogram of 1mb median windowed LD scores"))

  mean_ld_hist = ggplot(sub_df, aes(x=mean_r2))+
                    geom_histogram()+
                    labs(title = paste0("Data set: " , c,", genome-wide histogram of 1mb mean windowed LD scores"))
 
  print("write to file")  
  ggsave(paste0('external/ld_plots/cluster_', c,'_median_1mb_LD_histogram','.png'), median_ld_hist)
  ggsave(paste0('external/ld_plots/cluster_', c,'_mean_1mb_LD_histogram','.png'), mean_ld_hist)
  
  }



