############
# load libraries

library(tidyverse)
library(ggman)
theme_set(theme_bw())


############
# set path
data_path = "C:/Users/camnu/bin/salmon-genomic-vulnerability/data/"
setwd(data_path)


#downloaded in tabular format, with the unplaced contigs excluded
ssa_annotations = read_tsv("raw/SSA_genes.bed", col_names = c("chr", "bp_start", "bp_end", "ID"))
#and accompanying file
ssa_genes = read_tsv("raw/ssalar_gene_result.txt")
ssa_genes
names(ssa_genes)
ssa_genes$chr = unlist(lapply(ssa_genes$chromosome, function(x) as.integer(str_replace(x, "ssa", ""))))
ssa_genes = ssa_genes[!is.na(ssa_genes$chr),]
ssa_genes = ssa_genes[!is.na(ssa_genes$start_position_on_the_genomic_accession),]
ssa_genes = ssa_genes[!is.na(ssa_genes$end_position_on_the_genomic_accession),]

print("make the comparisons to ssa_genes, has the chromosomes, descriptions, IDs and positions")

#Tony says to use those ones because they have the postions that match the .map file
#whereas the ncbi one may have some small differences that could cause headaches.


################################################################################
################################################################################
print("load the PCA outliers")
#snp, position, p value
loading_full_info = read_tsv("interim/PCA_loadings_and_p_values_full_individual_set.tsv")
sig_pc_snps = loading_full_info[loading_full_info$p_values < 0.05,]
#18k sig snps, need to filter this down as well

#highly sig snps
sig_pc_snps = loading_full_info[loading_full_info$p_values < 0.00001,]
#then group them into location clusters


################################################################################
################################################################################
print("load the fst outliers")
#snp, position, fst, comparison_id
sig_fst_snps = read_tsv("interim/all_pw_fst_gte_15.tsv")
sig_fst_snps = sig_fst_snps[!is.na(sig_fst_snps$CHR),]
fst_summary = read_tsv("external/fst_plots/fst_summary_stats.tsv")
fst_summary = fst_summary[fst_summary$data_set == "full",]

print("TODO - come up with some additional filtering params, 12k is too many to check")
#possibly:
#    shared across all 3, 
#    spikes with x number of loci in a short distance
#    highest outliers
#    hisher threshold at 0.5 - then look for groups of SNPs in same regions

################################################################################
################################################################################
print("load the LD outliers")
#snp, position, LD, group
ld_df = read_tsv("raw/LD_files/ld_summary_output.tsv")

ld_df = drop_na(ld_df)
sig_LD_snps = ld_df %>%
                group_by(cluster) %>%
                  summarize(n = sum(n_loci), 
                            mean_mean_r2 = mean(mean_r2),
                            sd_mean_r2 = sd(mean_r2),
                            mean_med_r2 = mean(median_r2), 
                            sd_med_r2 = sd(median_r2))
  
sig_LD_snps$high_ld_threshold = unlist(lapply(1:nrow(sig_LD_snps), function(i){
  sig_LD_snps$mean_med_r2[[i]] + (4*sig_LD_snps$sd_med_r2[[i]])
}))


thresh_dict = sig_LD_snps$high_ld_threshold
names(thresh_dict) = sig_LD_snps$cluster

ld_df$is_high_ld = unlist(lapply(1:nrow(ld_df), function(i){
  ld_df$median_r2[[i]] > thresh_dict[[ld_df$cluster[[i]]]]
}))

print("high LD defined as median r2 for a window being  > 4 sd from the genome wide mean for the given cluster")
high_ld_df = ld_df[ld_df$is_high_ld,] #79 locations to inspect



################################################################################
################################################################################

print("add and ID for the windows")
high_ld_df$window_id = unlist(lapply(1:nrow(high_ld_df), function(i) paste0("sig_ld_window_", i)))

names(ssa_genes)

#chr =3   
#win_start = 4000001  
#win_end = 5000000
find_genes_in_window = function(win_id, chr, win_start, win_end, ssa_genes){
    
  sub_ssa_genes = ssa_genes[ssa_genes$chr == chr, ]  
  #less than the window end, so at least part of the front of the gene is in the window
  sub_ssa_genes = sub_ssa_genes[sub_ssa_genes$start_position_on_the_genomic_accession <= win_end,]
  #greater than the start, so that some of the tail hits the gene
  sub_ssa_genes = sub_ssa_genes[sub_ssa_genes$end_position_on_the_genomic_accession >= win_start,]
  
  sub_ssa_genes$window_id = win_id

  return(sub_ssa_genes)
}


ld_gene_matches = data.frame()

#i =1
for(i in 1:nrow(high_ld_df)){
  win_id = high_ld_df$window_id[[i]]

  chr = high_ld_df$chromosome[[i]]
  win_start = high_ld_df$win_start[[i]]
  win_end = high_ld_df$win_end[[i]]
  
  genes_in_window = find_genes_in_window(win_id, chr, win_start, win_end, ssa_genes)
  
  ld_gene_matches = rbind(ld_gene_matches, genes_in_window)
}


write_tsv(ld_gene_matches, "interim/sig_ld_gene_matches.tsv")


#snp_id = sig_fst_snps$SNP[[405]]
#chr = sig_fst_snps$CHR[[405]]
#pos = sig_fst_snps$POS[[405]]
#
find_snp_in_genes = function(snp_id, chr, pos, ssa_genes){
  
  sub_ssa_genes = ssa_genes[ssa_genes$chr == chr, ]  
  
  sub_ssa_genes = sub_ssa_genes[sub_ssa_genes$start_position_on_the_genomic_accession <= pos,]
  sub_ssa_genes = sub_ssa_genes[sub_ssa_genes$end_position_on_the_genomic_accession >= pos,]
  
  return(sub_ssa_genes)
  
}


fst_gene_matches = data.frame()

for(i in 1:nrow(sig_fst_snps)){
  gene_match = find_snp_in_genes(sig_fst_snps$SNP[[i]], sig_fst_snps$CHR[[i]], sig_fst_snps$POS[[i]], ssa_genes)
  if(nrow(gene_match) > 0){
    gene_match$snp_matched = sig_fst_snps$SNP[[i]]
    fst_gene_matches = rbind(fst_gene_matches, gene_match)
  }
}

fst_gene_matches
fst_gene_matches = fst_gene_matches[!is.na(fst_gene_matches$GeneID),]
write_tsv(fst_gene_matches, "interim/sig_fst_gene_matches.tsv")


pca_gene_matches = data.frame()

for(i in 1:nrow(sig_pc_snps)){
  gene_match = find_snp_in_genes(sig_pc_snps$snp[[i]], sig_pc_snps$chromosome[[i]], sig_pc_snps$physical_distance[[i]], ssa_genes)
  if(nrow(gene_match) > 0){
    gene_match$snp_matched = sig_pc_snps$snp[[i]]
    pca_gene_matches = rbind(pca_gene_matches, gene_match)
  }
}
pca_gene_matches
write_tsv(pca_gene_matches, "interim/sig_pca_gene_matches.tsv")


################################################################################
################################################################################
################################################################################
################################################################################
ld_gene_matches = read_tsv("interim/sig_ld_gene_matches.tsv")
fst_gene_matches = read_tsv("interim/sig_fst_gene_matches.tsv")
pca_gene_matches = read_tsv("interim/sig_pca_gene_matches.tsv")


print("inspecting the gene matches")
ld_gene_matches #high_ld_df
fst_gene_matches #high_ld_df, fst_summary 
pca_gene_matches #sig_pc_snps

print("look for concordance among the three, also whiddle down based off of other criteria")



matching_all_three = ld_gene_matches[ld_gene_matches$GeneID %in% fst_gene_matches$GeneID,]
matching_all_three = matching_all_three[matching_all_three$GeneID %in% pca_gene_matches$GeneID,]

table(matching_all_three$chromosome)
#10 regions popped up in all three things, on chr 11, 22, 23, and 26

write_tsv(matching_all_three, "interim/sig_all_three_tests.tsv")

fst_and_PCA_matches = pca_gene_matches[pca_gene_matches$GeneID %in% fst_gene_matches$GeneID,]
fst_and_PCA_matches #416 matches in both

length(unique(fst_and_PCA_matches$GeneID)) #239 unique genes matched to

write_tsv(fst_and_PCA_matches, "interim/sig_fst_and_pca_genes.tsv")


print("look for IDs of interest")

vgll3_id = 106586514
six6_id = 106610974
MHCII_idA = 106564142
MHCII_idB = 106564291

vgll3_id %in% fst_and_PCA_matches$GeneID #no
six6_id %in% fst_and_PCA_matches$GeneID   #no
MHCII_idA %in% fst_and_PCA_matches$GeneID #no
MHCII_idB %in% fst_and_PCA_matches$GeneID #no

vgll3_id %in% fst_gene_matches$GeneID #no
six6_id %in% fst_gene_matches$GeneID   #no
MHCII_idA %in% fst_gene_matches$GeneID #no
MHCII_idB %in% fst_gene_matches$GeneID #no

vgll3_id %in% pca_gene_matches$GeneID #no
six6_id %in% pca_gene_matches$GeneID   #YES
MHCII_idA %in% pca_gene_matches$GeneID #no
MHCII_idB %in% pca_gene_matches$GeneID #no



################################################################################
################################################################################

print("remember to check - are the NL individuals popping out in the lat/long corrected RDA the NS cluster from the southwest, or possibly admixed individuals?")








