############
# load libraries

library(tidyverse)
library(ggman)
theme_set(theme_bw())


############
# set path
data_path = "C:/Users/camnu/bin/salmon-genomic-vulnerability/data/"
setwd(data_path)


#downloaded in tabular format, with the unplaced contigs excluded
ssa_annotations = read_tsv("raw/SSA_genes.bed", col_names = c("chr", "bp_start", "bp_end", "ID"))
#and accompanying file
ssa_genes = read_tsv("raw/ssalar_gene_result.txt")

print("make the comparisons to ssa_genes, has the chromosomes, descriptions, IDs and positions")

#Tony says to use those ones because they have the postions that match the .map file
#whereas the ncbi one may have some small differences that could cause headaches.


################################################################################
################################################################################
print("load the PCA outliers")
#snp, position, p value
loading_full_info = read_tsv("interim/PCA_loadings_and_p_values_full_individual_set.tsv")
sig_pc_snps = loading_full_info[loading_full_info$p_values < 0.05,]
#18k sig snps, need to filter this down as well

#highly sig snps
sig_pc_snps = loading_full_info[loading_full_info$p_values < 0.00001,]
#then group them into location clusters


################################################################################
################################################################################
print("load the fst outliers")
#snp, position, fst, comparison_id
sig_fst_snps = read_tsv("interim/all_pw_fst_gte_15.tsv")
fst_summary = read_tsv("external/fst_plots/fst_summary_stats.tsv")
fst_summary = fst_summary[fst_summary$data_set == "full",]

print("TODO - come up with some additional filtering params, 12k is too many to check")
#possibly:
#    shared across all 3, 
#    spikes with x number of loci in a short distance
#    highest outliers
#    hisher threshold at 0.5 - then look for groups of SNPs in same regions

################################################################################
################################################################################
print("load the LD outliers")
#snp, position, LD, group
ld_df = read_tsv("raw/LD_files/ld_summary_output.tsv")

ld_df = drop_na(ld_df)
sig_LD_snps = ld_df %>%
                group_by(cluster) %>%
                  summarize(n = sum(n_loci), 
                            mean_mean_r2 = mean(mean_r2),
                            sd_mean_r2 = sd(mean_r2),
                            mean_med_r2 = mean(median_r2), 
                            sd_med_r2 = sd(median_r2))
  
sig_LD_snps$high_ld_threshold = unlist(lapply(1:nrow(sig_LD_snps), function(i){
  sig_LD_snps$mean_med_r2[[i]] + (4*sig_LD_snps$sd_med_r2[[i]])
}))


thresh_dict = sig_LD_snps$high_ld_threshold
names(thresh_dict) = sig_LD_snps$cluster

ld_df$is_high_ld = unlist(lapply(1:nrow(ld_df), function(i){
  ld_df$median_r2[[i]] > thresh_dict[[ld_df$cluster[[i]]]]
}))

print("high LD defined as median r2 for a window being  > 4 sd from the genome wide mean for the given cluster")
high_ld_df = ld_df[ld_df$is_high_ld,] #79 locations to inspect



################################################################################
################################################################################




print("remember to check - are the NL individuals popping out in the lat/long corrected RDA the NS cluster from the southwest, or possibly admixed individuals?")