############
# load libraries

library(tidyverse)
library(ggman)
theme_set(theme_bw())

############
# set path
data_path = "C:/Users/camnu/bin/salmon-genomic-vulnerability/data/"
setwd(data_path)

ssa_v3_gene_infile = "raw/genome_update/GCF_905237065.1/genomic.gtf"

cols = c("seqname", "source", "feature", "start", "end", "score", "strand", 
         "reading_frame", "attribute")

ssa_genes_v3 = read_tsv(ssa_v3_gene_infile, skip = 4, col_names = cols)

ssa_genes_v3 = ssa_genes_v3[ssa_genes_v3$feature == "gene",]

chr_dict = 1:29
names(chr_dict) = c("NC_059442.1", "NC_059443.1", "NC_059444.1", "NC_059445.1", "NC_059446.1", 
                    "NC_059447.1", "NC_059448.1", "NC_059449.1", "NC_059450.1", "NC_059451.1", 
                    "NC_059452.1", "NC_059453.1", "NC_059454.1", "NC_059455.1", "NC_059456.1", 
                    "NC_059457.1", "NC_059458.1", "NC_059459.1", "NC_059460.1", "NC_059461.1", 
                    "NC_059462.1", "NC_059463.1", "NC_059464.1", "NC_059465.1", "NC_059466.1", 
                    "NC_059467.1", "NC_059468.1", "NC_059469.1", "NC_059470.1")

ssa_genes_v3$chr = unlist(sapply(ssa_genes_v3$seqname, function(x){
                                  tryCatch(
                                    {chr_dict[[x]]}, 
                                  error = function(msg){
                                    return(NA)
                                  })}))

print("This is the genes from the transcripts to compare to the lists of sig SNPs")
table(ssa_genes_v3$chr)
ssa_genes_v3 = ssa_genes_v3[!is.na(ssa_genes_v3$chr),]

################################################################################
################################################################################
print("load the PCA outliers")
#snp, position, p value
loading_full_info = read_tsv("interim/PCA_loadings_and_p_values_full_individual_set.tsv")
sig_pc_snps = loading_full_info[loading_full_info$p_values < 0.05,]
#18k sig snps, need to filter this down as well

loading_full_info$bonf_P = p.adjust(loading_full_info$p_values, method= "bonferroni")

#highly sig snps
sig_pc_snps = loading_full_info[loading_full_info$bonf_P < 0.05,]
#then group them into location clusters


################################################################################
################################################################################
print("load the fst outliers")
#snp, position, fst, comparison_id
sig_fst_snps = read_tsv("interim/all_pw_fst_gte_15.tsv")
sig_fst_snps = sig_fst_snps[!is.na(sig_fst_snps$CHR),]
sig_fst_snps = sig_fst_snps[sig_fst_snps$CHR != 30,]

fst_summary = read_tsv("external/fst_plots/fst_summary_stats.tsv")
fst_summary = fst_summary[fst_summary$data_set == "full",]

print("TODO - come up with some additional filtering params, 12k is too many to check")
#possibly:
#    shared across all 3, 
#    spikes with x number of loci in a short distance
#    highest outliers
#    hisher threshold at 0.5 - then look for groups of SNPs in same regions

################################################################################
################################################################################
print("load the LD outliers")
#snp, position, LD, group
ld_df = read_tsv("raw/LD_files/new_noLDfilter_ld_summary_output.tsv")

ld_df = drop_na(ld_df)
sig_LD_snps = ld_df %>%
                group_by(cluster) %>%
                  summarize(n = sum(n_loci), 
                            mean_mean_r2 = mean(mean_r2),
                            sd_mean_r2 = sd(mean_r2),
                            mean_med_r2 = mean(median_r2), 
                            sd_med_r2 = sd(median_r2))
  
sig_LD_snps$high_ld_threshold = unlist(lapply(1:nrow(sig_LD_snps), function(i){
  sig_LD_snps$mean_med_r2[[i]] + (3*sig_LD_snps$sd_med_r2[[i]])
}))


thresh_dict = sig_LD_snps$high_ld_threshold
names(thresh_dict) = sig_LD_snps$cluster

ld_df$is_high_ld = unlist(lapply(1:nrow(ld_df), function(i){
  ld_df$median_r2[[i]] > thresh_dict[[ld_df$cluster[[i]]]]
}))

print("high LD defined as median r2 for a window being  > 3 sd from the genome wide mean for the given cluster")
high_ld_df = ld_df[ld_df$is_high_ld,] #31 locations to inspect
# only 19 with 5 or more comparisons as their basis
high_ld_df = high_ld_df[high_ld_df$n_comparisons >= 5,]

table(high_ld_df$cluster)
table(high_ld_df$cluster, high_ld_df$chromosome)

gte5_high_ld = high_ld_df[high_ld_df$n_comparisons >= 5,]
table(gte5_high_ld$cluster)
table(gte5_high_ld$cluster, gte5_high_ld$chromosome)


v_high_ld = high_ld_df[high_ld_df$median_r2 > 0.8,]

length(v_high_ld$win_start)
length(unique(v_high_ld$win_start))
table(v_high_ld$chromosome)

table(v_high_ld$cluster)
#full Newfoundland 
#26            5 

table(v_high_ld$cluster, v_high_ld$chromosome)

in_group = v_high_ld[v_high_ld$cluster != "full",]

table(in_group$cluster, in_group$chromosome)

################################################################################
################################################################################

print("add and ID for the windows")
high_ld_df$window_id = unlist(lapply(1:nrow(high_ld_df), function(i) paste0("sig_ld_window_", i)))


print("swap the ssa genes here")
#chr =3   
#win_start = 4000001  
#win_end = 5000000
find_genes_in_window = function(win_id, chr, win_start, win_end, ssa_genes_v3){
    
  sub_ssa_genes = ssa_genes_v3[ssa_genes_v3$chr == chr, ]  
  #less than the window end, so at least part of the front of the gene is in the window
  sub_ssa_genes = sub_ssa_genes[sub_ssa_genes$start <= win_end,]
  #greater than the start, so that some of the tail hits the gene
  sub_ssa_genes = sub_ssa_genes[sub_ssa_genes$end >= win_start,]
  
  sub_ssa_genes$window_id = win_id

  return(sub_ssa_genes)
}


ld_gene_matches = data.frame()

#i =1
for(i in 1:nrow(high_ld_df)){
  win_id = high_ld_df$window_id[[i]]

  chr = high_ld_df$chromosome[[i]]
  win_start = high_ld_df$win_start[[i]]
  win_end = high_ld_df$win_end[[i]]
  
  genes_in_window = find_genes_in_window(win_id, chr, win_start, win_end, ssa_genes_v3)
  
  ld_gene_matches = rbind(ld_gene_matches, genes_in_window)
}


write_tsv(ld_gene_matches, "interim/sig_ld_gene_matches.tsv")

high_n_ld_matches = data.frame()
gte5_high_ld
gte5_high_ld$window_id = unlist(lapply(1:nrow(gte5_high_ld), function(i) paste0("sig_ld_window_high_n", i)))

for(i in 1:nrow(gte5_high_ld)){
  win_id = gte5_high_ld$window_id[[i]]
  
  chr = gte5_high_ld$chromosome[[i]]
  win_start = gte5_high_ld$win_start[[i]]
  win_end = gte5_high_ld$win_end[[i]]
  
  genes_in_window = find_genes_in_window(win_id, chr, win_start, win_end, ssa_genes_v3)
  
  high_n_ld_matches = rbind(high_n_ld_matches, genes_in_window)
}

write_tsv(high_n_ld_matches, "interim/sig_ld_gene_matches_high_n.tsv")
write_tsv(gte5_high_ld, "interim/high_LD_high_n_regions.tsv")




#snp_id = sig_fst_snps$SNP[[405]]
#chr = sig_fst_snps$CHR[[405]]
#pos = sig_fst_snps$POS[[405]]
#
find_snp_in_genes = function(snp_id, chr, pos, ssa_genes_v3){
  
  sub_ssa_genes = ssa_genes_v3[ssa_genes_v3$chr == chr, ]  
  
  sub_ssa_genes = sub_ssa_genes[sub_ssa_genes$start <= pos,]
  sub_ssa_genes = sub_ssa_genes[sub_ssa_genes$end >= pos,]
  
  return(sub_ssa_genes)
  
}


fst_gene_matches = data.frame()

for(i in 1:nrow(sig_fst_snps)){
  gene_match = find_snp_in_genes(sig_fst_snps$SNP[[i]], sig_fst_snps$CHR[[i]], sig_fst_snps$POS[[i]], ssa_genes_v3)
  if(nrow(gene_match) > 0){
    gene_match$snp_matched = sig_fst_snps$SNP[[i]]
    fst_gene_matches = rbind(fst_gene_matches, gene_match)
  }
}

fst_gene_matches
fst_gene_matches = fst_gene_matches[!is.na(fst_gene_matches$attribute),]
write_tsv(fst_gene_matches, "interim/sig_fst_gene_matches.tsv")


pca_gene_matches = data.frame()

for(i in 1:nrow(sig_pc_snps)){
  gene_match = find_snp_in_genes(sig_pc_snps$snp[[i]], sig_pc_snps$chromosome[[i]], sig_pc_snps$physical_distance[[i]], ssa_genes_v3)
  if(nrow(gene_match) > 0){
    gene_match$snp_matched = sig_pc_snps$snp[[i]]
    pca_gene_matches = rbind(pca_gene_matches, gene_match)
  }
}
pca_gene_matches
write_tsv(pca_gene_matches, "interim/sig_pca_gene_matches.tsv")


################################################################################
################################################################################
################################################################################
################################################################################
ld_gene_matches = read_tsv("interim/sig_ld_gene_matches.tsv")
fst_gene_matches = read_tsv("interim/sig_fst_gene_matches.tsv")
pca_gene_matches = read_tsv("interim/sig_pca_gene_matches.tsv")


print("inspecting the gene matches")
ld_gene_matches #high_ld_df
fst_gene_matches #high_ld_df, fst_summary 
pca_gene_matches #sig_pc_snps

print("look for concordance among the three, also whiddle down based off of other criteria")

#matching_all_three = high_n_ld_matches[high_n_ld_matches$GeneID %in% fst_gene_matches$GeneID,]

matching_all_three = ld_gene_matches[ld_gene_matches$attribute  %in% fst_gene_matches$attribute ,]
matching_all_three = matching_all_three[matching_all_three$attribute  %in% pca_gene_matches$attribute ,]

table(matching_all_three$chromosome)
#no matches across all three
#write_tsv(matching_all_three, "interim/sig_all_three_tests.tsv")

fst_and_PCA_matches = pca_gene_matches[pca_gene_matches$attribute %in% fst_gene_matches$attribute,]
fst_and_PCA_matches #77 matches in both

fst_and_PCA_matches$matchtype = "fst_and_PCA"

length(unique(fst_and_PCA_matches$attribute)) #40 unique genes matched to

write_tsv(fst_and_PCA_matches, "interim/sig_fst_and_pca_genes.tsv")

fst_and_LD_matches = ld_gene_matches[ld_gene_matches$attribute  %in% fst_gene_matches$attribute ,]
fst_and_LD_matches #17 matches

fst_and_LD_matches$matchtype = "fst_and_LD_matches"

write_tsv(fst_and_PCA_matches, "interim/sig_fst_and_LD_genes.tsv")


PCA_and_LD_matches = ld_gene_matches[ld_gene_matches$attribute  %in% pca_gene_matches$attribute ,]
PCA_and_LD_matches #2 matches

PCA_and_LD_matches$matchtype = "PCA_and_LD_matches"

write_tsv(PCA_and_LD_matches, "interim/PCA_and_LD_genes.tsv")


print("parse the attributes to see what is there")

#att = PCA_and_LD_matches$attribute[[1]]
#att = PCA_and_LD_matches$attribute[[2]]
parse_attributes = function(att){
  
  att = gsub( '"', '', att)
  att_fields = strsplit(att, ";")[[1]]
  
  att_vals = sapply(att_fields, function(x) {
      z = strsplit(x, " ")[[1]]
      z[[length(z)]]})
  
  outdat = c(att_vals[[1]], gsub("GeneID:", "", att_vals[[3]]), 
                att_vals[[5]])
  names(outdat) = c("Gene", "GeneID","gene_biotype")
  
  return(outdat)
}


#' Take the attributes column, parse the information and add it in as new columns
attributes_to_df = function(df){
  parsed_attr = do.call(rbind, lapply(df$attribute, parse_attributes))
  out_df = cbind(df, parsed_attr)
  out_df$Gene = unlist(out_df$Gene)
  out_df$GeneID = unlist(out_df$GeneID)
  out_df$gene_biotype = unlist(out_df$gene_biotype)
  return(out_df)
}


PCA_and_LD_matches = attributes_to_df(PCA_and_LD_matches)
fst_and_PCA_matches = attributes_to_df(fst_and_PCA_matches)
fst_and_LD_matches = attributes_to_df(fst_and_LD_matches)


fst_gene_matches = attributes_to_df(fst_gene_matches)
pca_gene_matches = attributes_to_df(pca_gene_matches)
ld_gene_matches = attributes_to_df(ld_gene_matches)

print("look for IDs of interest")
print("TODO - get these for the new genome, or are they the same?")
vgll3_id = "106586514"
six6_id = "106610974"
MHCII_idA = "106564142"
MHCII_idB = "106564291"

vgll3_id %in% fst_and_PCA_matches$GeneID #no
six6_id %in% fst_and_PCA_matches$GeneID   #no
MHCII_idA %in% fst_and_PCA_matches$GeneID #no
MHCII_idB %in% fst_and_PCA_matches$GeneID #no

vgll3_id %in% fst_gene_matches$GeneID #no
six6_id %in% fst_gene_matches$GeneID   #no
MHCII_idA %in% fst_gene_matches$GeneID #no
MHCII_idB %in% fst_gene_matches$GeneID #no

vgll3_id %in% pca_gene_matches$GeneID #no
six6_id %in% pca_gene_matches$GeneID   #no
MHCII_idA %in% pca_gene_matches$GeneID #no
MHCII_idB %in% pca_gene_matches$GeneID #no

vgll3_id %in% ld_gene_matches$GeneID #no
six6_id %in% ld_gene_matches$GeneID   #no
MHCII_idA %in% ld_gene_matches$GeneID #no
MHCII_idB %in% ld_gene_matches$GeneID #no

fst_and_PCA_matches$window_id = NA
PCA_and_LD_matches$snp_matched = NA
fst_and_LD_matches$snp_matched = NA
all_combos_df = rbind(PCA_and_LD_matches, fst_and_PCA_matches)
all_combos_df = rbind(all_combos_df, fst_and_LD_matches)
all_combos_df = all_combos_df[, !duplicated(colnames(all_combos_df))]

all_combos_df = as_tibble(all_combos_df)

write_tsv(all_combos_df, "interim/salmon_PCA_fst_LD_genes_of_interest.tsv")

################################################################################
################################################################################

print("remember to check - are the NL individuals popping out in the lat/long corrected RDA the NS cluster from the southwest, or possibly admixed individuals?")



#the number of sig genes from the above, each location only counted once (57 unique genes)
#chr_num
#1  2  3  4  5  6  7  8  9 10 11 13 14 15 17 18 19 21 22 23 25 26 27 28 
#8  3  1  8  1  1  2  1  3  1  2  3  4  1  1  1  2  2  1  3  1  3  1  3 



