############
# load libraries

library(tidyverse)
library(ggman)
theme_set(theme_bw())
library(pcadapt)
library(data.table)




############
# set path
data_path = "C:/Users/camnu/bin/salmon-genomic-vulnerability/data/"
setwd(data_path)



############
print("generate the lists of individuals for creating the subsets for the PCAs")
print("have the full sets, remaking with just the high Q individuals to see if admixture affects the clustering")

full_per_individ_q_vals = read_tsv("interim/cluster_and_q_data_combined.tsv")

keep =  c("SiteCode", "fish_id", "location_id", "location_name", "Province" , "lat" , "lon" , "date_of_genotyping",
          "IID",  "CID", "cluster_name",  "Q1", "Q2", "Q3", "home_Q", "total_admixed_Q")  
sub_meta = full_per_individ_q_vals[keep]

per_individ_q_vals = full_per_individ_q_vals[full_per_individ_q_vals$home_Q >= 0.8,]


cluster_names = c("Labrador", "Newfoundland", "St. Lawrence")

#c = cluster_names[[3]]
#sub_df = full_per_individ_q_vals[full_per_individ_q_vals$cluster_name == c,]
#length(table(sub_df$location_name))

#c = "Labrador"
for(c in cluster_names){
  
  sub_df = per_individ_q_vals[per_individ_q_vals$cluster_name == c,]
  
  out_df = sub_df[c("SiteCode", "fish_id")]
  
  names(out_df) = c("FID", "IID")
  
  
  write_tsv(out_df, paste0("raw/", c, "_highQ_individual_list.tsv"))  
}


#347 samples:
#plink --bfile wildNA_220k_filtered --chr-set 30 no-xy --maf 0.01 --make-bed --keep Labrador_highQ_individual_list.tsv --out high_Q_Labrador_220k_filtered
#plink --bfile high_Q_Labrador_220k_filtered --chr-set 30 no-xy --recode --tab --out high_Q_Labrador_220k_filtered

#2153 samples:
#plink --bfile wildNA_220k_filtered --chr-set 30 no-xy --maf 0.01 --make-bed --keep Newfoundland_highQ_individual_list.tsv  --out high_Q_Newfoundland_220k_filtered
#plink --bfile high_Q_Newfoundland_220k_filtered --chr-set 30 no-xy --recode --tab --out high_Q_Newfoundland_220k_filtered

#1361 samples:
#plink --bfile wildNA_220k_filtered --chr-set 30 no-xy --maf 0.01 --make-bed --keep STLawrence_highQ_individual_list.tsv  --out high_Q_STLawrence_220k_filtered
#plink --bfile high_Q_STLawrence_220k_filtered --chr-set 30 no-xy --recode --tab --out high_Q_STLawrence_220k_filtered


list_of_high_q_bed_files = c('raw/high_Q_Labrador_220k_filtered.bed', 'raw/high_Q_Newfoundland_220k_filtered.bed', 'raw/high_Q_STLawrence_220k_filtered.bed')
names(list_of_high_q_bed_files) = c("LAB", "NFL", "STL")

list_of_high_q_fam_files = c('raw/high_Q_Labrador_220k_filtered.fam', 'raw/high_Q_Newfoundland_220k_filtered.fam', 'raw/high_Q_STLawrence_220k_filtered.fam')
names(list_of_high_q_fam_files) = c("LAB", "NFL", "STL")

list_of_high_q_map_files = c('raw/high_Q_Labrador_220k_filtered.map', 'raw/high_Q_Newfoundland_220k_filtered.map', 'raw/high_Q_STLawrence_220k_filtered.map')
names(list_of_high_q_map_files) = c("LAB", "NFL", "STL")


list_of_full_bed_files = c('raw/LD_files/Labrador_220k_filtered.bed' , 'raw/LD_files/Newfoundland_220k_filtered.bed', 'raw/LD_files/STLawrence_220k_filtered.bed')
names(list_of_full_bed_files) = c("LAB", "NFL", "STL")

list_of_full_fam_files = c('raw/LD_files/Labrador_220k_filtered.fam' , 'raw/LD_files/Newfoundland_220k_filtered.fam', 'raw/LD_files/STLawrence_220k_filtered.fam')
names(list_of_full_fam_files) = c("LAB", "NFL", "STL")

list_of_full_map_files = c('raw/LD_files/Labrador_220k_filtered.map' , 'raw/LD_files/Newfoundland_220k_filtered.map', 'raw/LD_files/STLawrence_220k_filtered.map')
names(list_of_full_map_files) = c("LAB", "NFL", "STL")


#c = "LAB"
for(c in names(list_of_full_bed_files)){
  print(paste0("on cluster: ", c))
  high_q_bed = list_of_high_q_bed_files[[c]]
  full_bed = list_of_full_bed_files[[c]]
  
  high_q_fam = list_of_high_q_fam_files[[c]]
  full_fam = list_of_full_fam_files[[c]]

  high_q_map = list_of_high_q_map_files[[c]]
  full_q_map = list_of_full_map_files[[c]]
  
  high_cluster_pca = read.pcadapt(high_q_bed, type = "bed")
  high_cluster_PCA_out = pcadapt(high_cluster_pca, K = 2, min.maf = 0.01)
  print("high Q only axis values:")
  print((high_cluster_PCA_out$singular.values)^2*100)   
  
  highQ_cluster_pca_Scores = data.frame(high_cluster_PCA_out$scores)
  colnames(highQ_cluster_pca_Scores) = c("PC1", "PC2")
  highQ_fam_df <- fread(high_q_fam) %>% select(V1, V2)
  colnames(highQ_fam_df) <- c("SiteCode", "fish_id")
  highQ_PCs_with_ids = cbind(highQ_fam_df, highQ_cluster_pca_Scores) 
  
  out_high_Q = left_join(highQ_PCs_with_ids, sub_meta)
  
  write_tsv(out_high_Q, paste0("interim/", c,"_high_Q_PCA_and_meta.tsv"))
  
  #get the pc loadings and p values as well
  high_map = read_tsv(high_q_map, col_names = c("chromosome", "snp" , "genetic_distance", "physical_distance"))
  highQ_loading_df = data.frame(high_cluster_PCA_out$loadings)
  names(highQ_loading_df)=c("loadings_PC1", "loadings_PC2")
  highQ_loading_full_info = as_tibble(cbind(high_map, highQ_loading_df))
  highQ_loading_full_info = as_tibble(cbind(highQ_loading_full_info, data.frame(p_values = high_cluster_PCA_out$pvalues)))
  write_tsv(highQ_loading_full_info, paste0("interim/", c,"_high_Q_PCA_loadings_pvals.tsv"))
  
    
  full_cluster_pca = read.pcadapt(full_bed, type = "bed")
  full_cluster_pca_out = pcadapt(full_cluster_pca, K = 2, min.maf = 0.01)
  print("full cluster axis values:")
  print((full_cluster_pca_out$singular.values)^2*100) 
  
  
  #add the graphing and output generation in here.
  
  full_cluster_pca_Scores = data.frame(full_cluster_pca_out$scores)
  colnames(full_cluster_pca_Scores) = c("PC1", "PC2")
  full_fam_df <- fread(full_fam) %>% select(V1, V2)
  colnames(full_fam_df) <- c("SiteCode", "fish_id")
  full_PCs_with_ids = cbind(full_fam_df, full_cluster_pca_Scores) 
  
  out_full = left_join(full_PCs_with_ids, sub_meta)

  write_tsv(out_full, paste0("interim/", c,"_full_indiv_PCA_and_meta.tsv"))
  

  #get the pc loadings and p values as well
  full_map = read_tsv(full_q_map, col_names = c("chromosome", "snp" , "genetic_distance", "physical_distance"))
  full_loading_df = data.frame(full_cluster_pca_out$loadings)
  names(full_loading_df)=c("loadings_PC1", "loadings_PC2")
  full_loading_full_info = as_tibble(cbind(full_map, full_loading_df))
  full_loading_full_info = as_tibble(cbind(full_loading_full_info, data.frame(p_values = full_cluster_pca_out$pvalues)))
  write_tsv(full_loading_full_info, paste0("interim/", c,"_full_PCA_loadings_pvals.tsv"))

}
# 
# [1] "on cluster: LAB"
# [1] "high Q only axis values:"
# [1] 3.104323 2.389208
# Joining, by = c("SiteCode", "fish_id")
# [1] "full cluster axis values:"                                                                                                                  
# [1] 2.811970 2.020818
# Joining, by = c("SiteCode", "fish_id")
# [1] "on cluster: NFL"                                                                                                                            
# [1] "high Q only axis values:"
# [1] 1.457462 1.197040
# Joining, by = c("SiteCode", "fish_id")
# [1] "full cluster axis values:"                                                                                                                  
# [1] 1.312261 1.092917
# Joining, by = c("SiteCode", "fish_id")
# [1] "on cluster: STL"                                                                                                                            
# [1] "high Q only axis values:"
# [1] 2.220433 1.847572
# Joining, by = c("SiteCode", "fish_id")
# [1] "full cluster axis values:"                                                                                                                  
# [1] 2.003713 1.601749
# Joining, by = c("SiteCode", "fish_id")


print("read in the cluster pca results, generate the PCA plots and viz with different colourings")

highQ_pca_out_file_list = c("interim/LAB_high_Q_PCA_and_meta.tsv", "interim/NFL_high_Q_PCA_and_meta.tsv", "interim/STL_high_Q_PCA_and_meta.tsv")
names(highQ_pca_out_file_list) = c("LAB", "NFL", "STL")

full_pca_out_file_list = c("interim/LAB_full_indiv_PCA_and_meta.tsv", "interim/NFL_full_indiv_PCA_and_meta.tsv", "interim/STL_full_indiv_PCA_and_meta.tsv")
names(full_pca_out_file_list) = c("LAB", "NFL", "STL")


print("turn the numbers above into axis labels and apply them to the plots as you go.")
high_pc1_scores = c(3.1, 1.5, 2.2)
names(high_pc1_scores) = c("LAB", "NFL", "STL")

high_pc2_scores = c(2.4, 1.2, 1.8)
names(high_pc2_scores) = c("LAB", "NFL", "STL")

full_pc1_scores = c(2.8, 1.3, 2.0)
names(full_pc1_scores) = c("LAB", "NFL", "STL")

full_pc2_scores = c(2.2, 1.1 ,1.6)
names(full_pc2_scores) = c("LAB", "NFL", "STL")

#c = "STL"
#c = "LAB"
for(c in names(highQ_pca_out_file_list)){

  high_q_pca = read_tsv(highQ_pca_out_file_list[[c]])
  
  full_pca = read_tsv(full_pca_out_file_list[[c]])
  
  
  highQ_pca_of_filtered_env = ggplot() + 
    geom_point(data = high_q_pca, aes(x = PC1, y = PC2, colour = as.factor(location_id)))  +
    labs(title = paste0("PCA of Environmental data, cluster: " , c),
         subtitle = " Individuals from cluster with admix > 0.8")+
    xlab(paste0("PC1 (",high_pc1_scores[[c]] , "% of variance explained)")) +
    ylab(paste0("PC2 (", high_pc2_scores[[c]], "% of variance explained)")) +
    theme(legend.position = "none")
  
  ggsave(paste0("interim/pca_plot_cluster_", c, "_high_Q_by_location.png"), highQ_pca_of_filtered_env)
  
  
  highQ_pvalues_loading_plot = ggman(high_q_pca, snp = "snp", 
                                             bp = "physical_distance", chrom = "chromosome", 
                                             pvalue = "p_values",
                                             logTransform = TRUE)+
                              labs(title = paste0("PCA P values, cluster: ", c), 
                                   subtitle = " Individuals from cluster with admix > 0.8 ")+
                              xlab("Chromosome") +
                              ylab("-log10(pvalue) value)")+
                              theme(panel.grid.major = element_blank(), 
                                    panel.grid.minor = element_blank())
  
  #  geom_hline(yintercept=0.15, linetype="dashed", colour="red", 1) + theme_minimal()+
  
  
  ggsave('external/full_individual_PCA_p_values.png', pvalues_loading_plot,  width = 8, height = 3)
  
  
  
  
  
  full_pca_of_filtered_env = ggplot() + 
    geom_point(data = full_pca, aes(x = PC1, y = PC2, colour = as.factor(location_id)))  +
    theme(legend.position = "none")+
    labs(title = paste0("PCA of Environmental data, cluster: " , c),
         subtitle = " Full set of individuals assigned to cluster")+
    xlab(paste0("PC1 (",full_pc1_scores[[c]] , "% of variance explained)")) +
    ylab(paste0("PC2 (", full_pc2_scores[[c]], "% of variance explained)")) +
    theme(legend.position = "none")
  
  ggsave(paste0("interim/pca_plot_cluster_", c, "_full_cluster_by_location.png"), full_pca_of_filtered_env)
  
  
}


highQ_pca_loadings_file_list = c("interim/LAB_high_Q_PCA_loadings_pvals.tsv", "interim/NFL_high_Q_PCA_loadings_pvals.tsv", "interim/STL_high_Q_PCA_loadings_pvals.tsv")
names(highQ_pca_loadings_file_list) = c("LAB", "NFL", "STL")

full_pca_loadings_file_list = c("interim/LAB_full_PCA_loadings_pvals.tsv", "interim/NFL_full_PCA_loadings_pvals.tsv", "interim/STL_full_PCA_loadings_pvals.tsv")
names(full_pca_loadings_file_list) = c("LAB", "NFL", "STL")


#c = "STL"
#c = "LAB"
for(c in names(highQ_pca_loadings_file_list)){
  
  high_q_pca = read_tsv(highQ_pca_loadings_file_list[[c]])
  
  full_pca = read_tsv(full_pca_loadings_file_list[[c]])
  
  
  
  highQ_pvalues_loading_plot = ggman(high_q_pca, snp = "snp", 
                                     bp = "physical_distance", chrom = "chromosome", 
                                     pvalue = "p_values",
                                     logTransform = TRUE)+
    labs(title = paste0("PCA P values, cluster: ", c), 
         subtitle = " Individuals from cluster with admix > 0.8 ")+
    xlab("Chromosome") +
    ylab("-log10(pvalue) value)")+
    theme(panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank())
  
  #  geom_hline(yintercept=0.15, linetype="dashed", colour="red", 1) + theme_minimal()+
  
  
  ggsave(paste0('interim/cluster_', c, '_highQ_individual_PCA_p_values.png'), highQ_pvalues_loading_plot,  width = 8, height = 3)
  
  
  
  

}










print("follow up idea")
# is this evidence of three different geographic variants?
# plot each of the groups on a map and use the PC1 of the intra cluster to colour, 
# do they seem to sort along a graident
# conversely could do an LM of lat/long with the PCs as the response.

#essentially may be seeing evidence of three genetic clusters, and within each of the 
#clusters there is evidence of adaptive geographic gradient
# supports other findings of heirarchical structing of salmon populations,  
# and can assert that the top level structure may be neutral differences arising from
# background in which the pops were separated in different glacial refugia




