import gc
import pandas as pd



if __name__ == "__main__":


    """ read in the raw genotype data file, the list of individuals, 
        and the map file with the new SNP genome order."""


    #the raw file with the genotypes for each individual, dosage encoded
    #snp_data = pd.read_csv("../data/raw/Salmo_220K_Merged2022_COMPLETE_ENVmatchyear.raw", delim_whitespace=True)      
    snp_data = pd.read_csv("../data/raw/demo.raw", sep= " ")      

    snp_data.head()
    gc.collect()


    #the map file with the genome order
    snp_map = pd.read_csv("../data/raw/genome_update/Ssa220K_v3_Genome_with_approx.map", sep = "\t")

    #the cluster and river assignments of individuals
    pop_metadata = pd.read_csv("../data/interim/full_data_PCA_kmeans_cluster_info.tsv", sep = "\t")

    #repeat on these two grouping scales:
    pop_metadata.Cluster
    pop_metadata.location_id




pop_metadata.columns


    """ for each of the three populations, and each river:
        obtain the major and minor allele counts for each marker, 
        store in a dict"""


    """ generate two output files, one for the populations, and one
        for the rivers.
            
            pop1 pop2 pop3 pop4
            5,1 1,1 4,0 0,4
            3,3 0,2 2,2 0,4
            1,5 0,2 2,2 1,3
        
        where each row is a SNP and the rows are ordered 
        as per the SNP's genome order        
    """