


#########################
# libraries
library(tidyverse)

library(data.table)

#install.packages("ggforce")
library(ggforce)
library(ggplot2)
library(rnaturalearth)
library(rnaturalearthdata)


############
# set path
data_path = "C:/Users/camnu/bin/salmon-genomic-vulnerability/data/"
setwd(data_path)

alter_cluster_name = function(x){
  if(x == "St. Lawrence"){
    return("Maritimes")
  }else{
    return(x)
  }
}

alter_cluster_IDs = function(x){
  if(x == "STL"){
    return("Maritimes")
  }else if (x == "NFL"){
    return("Newfoundland")
  }else if (x == "LAB"){
    return("Labrador")
  }else{
    return(x)
  }
}




#
# using a set of admixture sub runs to quantify the effect of European ancestry on the NFL/LAB differences
# 


##########
# 1. get list of significant fst outliers between NFL and LAB
#   - these are the regions of the genome that are different
# 

full_fst_file = "interim/fst_files/LAB_v_NFL_fst_full_pairwise_updated_genome.fst"

full_fst = read_tsv(full_fst_file)
mean_fst = mean(full_fst$FST, na.rm = TRUE)
sd_fst = sd(full_fst$FST, na.rm = TRUE)

genome_thresh = mean_fst + (3*sd_fst)

full_fst$high_fst = unlist(lapply(1:nrow(full_fst), function(i){
  full_fst$FST[[i]] > genome_thresh
}))

high_fst = full_fst[full_fst$high_fst==TRUE,]
high_fst #1648 markers with Fst score > 3sd from the mean for the pairwise comparion
print("1648 markers with Fst score > 3sd from the mean for the pairwise comparion")
print("2.9% on chr 1, 4.3 % on chr 23")
chr_table = (table(high_fst$CHR) / table(full_fst$CHR)) * 100

chr_table
sort(chr_table)
freq_chi = chisq.test(chr_table)
freq_chi
#not significant difference


#read in the full snp list, then subset everything but the above to make an exclude list.
full_snps = read_tsv("C:/Users/camnu/bin/salmon-euro-introgression/data/May2022_SNPchip_220K/SNPchip_100MS_SNPpanel_King7_Compare/SNPchip_220K/all_fish_Salmo220K_Polyhigh_maf01.map", 
                        col_names = c("CHR", "SNP", "MAP", "POS"))

exclude_snps = full_snps[!full_snps$SNP %in% high_fst$SNP, ]


write.table(exclude_snps$SNP, "Fst_LAB_NFL_exclude.tsv", col.names = F, row.names = F, quote = F, sep = "\t")


##########
# 2. Get the list of fish in the NFL and LAB and EU populations
#   - use new and old metadata, these are the fish we will use to analyze admixture
# 


clusters_per_location  = read_tsv("interim/cluster_and_q_data_combined.tsv")
clusters_per_location$Population = unlist(lapply(clusters_per_location$river_majority_cluster, alter_cluster_IDs))
clusters_per_location$location_id = clusters_per_location$SiteCode


nfl_and_lab = clusters_per_location[clusters_per_location$home_cluster != "STL",]

full_admix = read_tsv("C:/Users/camnu/bin/salmon-euro-introgression/data/May2022_SNPchip_220K/SNPchip_100MS_SNPpanel_King7_Compare/SNPchip_220K/all_data_admix_vals.tsv")
table(full_admix$data_class)
nor_fish = full_admix[full_admix$data_class == "Norway",]


print("write the combined list to a file")

out_df = rbind(nfl_and_lab[c("location_id", "fish_id")], nor_fish[c("location_id", "fish_id")])

names(out_df) = c("FID", "IID")
write_tsv(out_df, "Nor_NFL_LAB_individual_list.tsv")

base_out = nfl_and_lab[c("location_id", "fish_id")]
names(base_out) = c("FID", "IID")

write_tsv(base_out, "NFL_LAB_individual_list.tsv")

print("Admixture with:")
print("806 Norwegian")
print("431 LAB")
print("3244 NFL")

##########
# 3. use plink to subset the old bed/bim file (complete data set from SalmonEuAdmix)
#   - pull just the Fst outliers and the LAB/NFL/Eu fish
#   - pull just the Fst outleris and the LAB/NFL fish
# 


#plink --bfile /mnt/c/Users/camnu/bin/salmon-euro-introgression/data/May2022_SNPchip_220K/SNPchip_100MS_SNPpanel_King7_Compare/SNPchip_220K/wild_Salmo220K_NA_EU_LE_filtered --keep Nor_NFL_LAB_individual_list.tsv --exclude Fst_LAB_NFL_exclude.tsv --chr-set 30 no-xy --make-bed --out Nor_NFL_LAB_220K_filtered
## worked: 1634 variants and 4481 samples pass filters and QC.

#non LD filtered
#plink --bfile /mnt/c/Users/camnu/bin/salmon-euro-introgression/data/May2022_SNPchip_220K/SNPchip_100MS_SNPpanel_King7_Compare/SNPchip_220K/wild_Salmo220K_NA_EU_LE_filtered --keep Nor_NFL_LAB_individual_list.tsv --chr-set 30 no-xy --make-bed --out genome_wide_Nor_NFL_LAB_220K_filtered

#plink --bfile /mnt/c/Users/camnu/bin/salmon-euro-introgression/data/May2022_SNPchip_220K/SNPchip_100MS_SNPpanel_King7_Compare/SNPchip_220K/wild_Salmo220K_NA_EU_LE_filtered --keep NFL_LAB_individual_list.tsv --exclude Fst_LAB_NFL_exclude.tsv --chr-set 30 no-xy --make-bed --out NFL_LAB_220K_filtered
## worked: 1634 variants and 3675 samples pass filters and QC.

#plink --bfile /mnt/c/Users/camnu/bin/salmon-euro-introgression/data/May2022_SNPchip_220K/SNPchip_100MS_SNPpanel_King7_Compare/SNPchip_220K/wild_Salmo220K_NA_EU_LE_filtered --keep NFL_LAB_individual_list.tsv --chr-set 30 no-xy --make-bed --out genome_wide_NFL_LAB_220K_filtered

##########
# 4. run admixture
# 
# - k=2 and k=3 -> Fst outliers and the LAB/NFL/Eu fish ()
#     - if the admixture proprotions are really high, then can say the regions of European contact are driving the differentiation between the groups
#     - if admixture proportions are low, can say that the Eu admixture is not in the differentiated parts of the genome, likely shared Eu ancestry.
# - k=2         ->  Fst outleris and the LAB/NFL fish (serves as a baseline)



# admixture Nor_NFL_LAB_220K_filtered.bed 2
# admixture Nor_NFL_LAB_220K_filtered.bed 3
# 

# admixture genome_wide_Nor_NFL_LAB_220K_filtered.bed 2
# admixture genome_wide_Nor_NFL_LAB_220K_filtered.bed 3

# admixture NFL_LAB_220K_filtered.bed 2
# admixture genome_wide_NFL_LAB_220K_filtered 2


##########
#5. load the admixture outputs, plot and analyze the results

genome_wide_Eu_NA_AdmixQ2_data = fread("genome_wide_Nor_NFL_LAB_220K_filtered.2.Q")
colnames(genome_wide_Eu_NA_AdmixQ2_data) = c("Q1", "Q2")
genome_wide_Eu_NA_AdmixQ3_data = fread("genome_wide_Nor_NFL_LAB_220K_filtered.3.Q")
colnames(genome_wide_Eu_NA_AdmixQ3_data) = c("Q1", "Q2", "Q3")
genome_wide_NA_AdmixQ2_data = fread("genome_wide_NFL_LAB_220K_filtered.2.Q")
colnames(genome_wide_NA_AdmixQ2_data) = c("Q1", "Q2")

Eu_NA_AdmixQ2_data = fread("Nor_NFL_LAB_220K_filtered.2.Q")
colnames(Eu_NA_AdmixQ2_data) = c("Q1", "Q2")
Eu_NA_AdmixQ3_data = fread("Nor_NFL_LAB_220K_filtered.3.Q")
colnames(Eu_NA_AdmixQ3_data) = c("Q1", "Q2", "Q3")
NA_AdmixQ2_data = fread("NFL_LAB_220K_filtered.2.Q")
colnames(NA_AdmixQ2_data) = c("Q1", "Q2")



print("the fam files are in alpha order - need to match that!")
nor_fish$home_cluster = "NOR"
#out_df = rbind(nfl_and_lab[c("location_id", "fish_id")], nor_fish[c("location_id", "fish_id")])
base_EU_NA_df = rbind(nfl_and_lab[c("location_id", "fish_id", "home_cluster")], nor_fish[c("location_id", "fish_id", "home_cluster")])
base_EU_NA_df = base_EU_NA_df[order(base_EU_NA_df$location_id, base_EU_NA_df$fish_id),]

nfl_lab_df = nfl_and_lab[c("location_id", "fish_id", "home_cluster")]
nfl_lab_df = nfl_lab_df[order(nfl_lab_df$location_id, nfl_lab_df$fish_id),]

genome_wide_EU_NA_df_Q2 = cbind(base_EU_NA_df, genome_wide_Eu_NA_AdmixQ2_data)
genome_wide_EU_NA_df_Q3 = cbind(base_EU_NA_df, genome_wide_Eu_NA_AdmixQ3_data)
genome_wide_base_NA_df = cbind(nfl_lab_df, genome_wide_NA_AdmixQ2_data)

EU_NA_df_Q2 = cbind(base_EU_NA_df, Eu_NA_AdmixQ2_data)
EU_NA_df_Q3 = cbind(base_EU_NA_df, Eu_NA_AdmixQ3_data)
base_NA_df = cbind(nfl_lab_df, NA_AdmixQ2_data)


genome_wide_Eu_per_group_Q2 = genome_wide_EU_NA_df_Q2 %>%
  group_by(home_cluster) %>%
  summarize(mean_Q1 = mean(Q1), 
            mean_q2 = mean(Q2),
            count = n())
genome_wide_Eu_per_group_Q2

#     home_cluster mean_Q1 mean_q2 count
#   1 LAB           0.987  0.0131    431
#   2 NFL           0.996  0.00421  3244
#   3 NOR           0.0179 0.982     806


genome_wide_Eu_per_group_Q3 = genome_wide_EU_NA_df_Q3 %>%
  group_by(home_cluster) %>%
  summarize(mean_Q1 = mean(Q1), 
            mean_q2 = mean(Q2),
            mean_q3 = mean(Q3),
            count = n())
genome_wide_Eu_per_group_Q3

#     home_cluster mean_Q1 mean_q2  mean_q3 count
#   1 LAB           0.917  0.0819  0.000807   431
#   2 NFL           0.126  0.871   0.00323   3244
#   3 NOR           0.0130 0.00838 0.979      806


genome_wide_base_NA_per_group_Q2 = genome_wide_base_NA_df %>%
  group_by(home_cluster) %>%
  summarize(mean_Q1 = mean(Q1), 
            mean_q2 = mean(Q2),
            count = n())

genome_wide_base_NA_per_group_Q2
#   home_cluster mean_Q1 mean_q2 count
# 1 LAB           0.0829   0.917   431
# 2 NFL           0.869    0.131  3244
#################################
Eu_per_group_Q2 = EU_NA_df_Q2 %>%
                    group_by(home_cluster) %>%
                            summarize(mean_Q1 = mean(Q1), 
                                      mean_q2 = mean(Q2),
                                      count = n())
Eu_per_group_Q2

#     home_cluster mean_Q1 mean_q2 count
#   1 LAB           0.413   0.587    431
#   2 NFL           0.976   0.0244  3244
#   3 NOR           0.0435  0.956    806

Eu_per_group_Q3 = EU_NA_df_Q3 %>%
                    group_by(home_cluster) %>%
                        summarize(mean_Q1 = mean(Q1), 
                                  mean_q2 = mean(Q2),
                                  mean_q3 = mean(Q3),
                                  count = n())
Eu_per_group_Q3

#     home_cluster mean_Q1 mean_q2 mean_q3 count
#   1 LAB           0.141   0.801  0.0583    431
#   2 NFL           0.948   0.0469 0.00554  3244
#   3 NOR           0.0358  0.0297 0.934     806

base_NA_per_group_Q2 = base_NA_df %>%
                        group_by(home_cluster) %>%
                        summarize(mean_Q1 = mean(Q1), 
                                  mean_q2 = mean(Q2),
                                  count = n())
base_NA_per_group_Q2

#     home_cluster mean_Q1 mean_q2 count
#   1 LAB            0.154  0.846    431
#   2 NFL            0.951  0.0488  3244



### 
# RESULTS
# - the FST SNPs separate out the LAB and NFL with high fidelity
# - the FST SNPS SEPARATE out the EU and NA fish well
# - The genome wide SNPS ____ separate out the EU and NA fish welll
# - therefore the regions of the genome of high differentation between the NFL and the LAB populations are
#      and we can conclude that


#7c suggests that the FST separating the LAB from the NFL are the European admixture SNPS

# Generate supporting plots to this effect:
# make an additional supplementary

######
# Figure S7 - admixture of LAB-NFL-EU
Q_to_colour = function(Q){  
  col_dict = c("blue", "red", "yellow")  
  names(col_dict)  =  c("Q1", "Q2", "Q3")
  return(col_dict[[Q]])
}

#A)  LAB-NFL-EU - GENOME WIDE
genome_wide_EU_NA_df_Q2
genome_wide_Eu_per_group_df_qlong = gather(genome_wide_EU_NA_df_Q2, Q_type, Q, Q1, Q2) 
genome_wide_Eu_per_group_df_qlong = genome_wide_Eu_per_group_df_qlong[order(genome_wide_Eu_per_group_df_qlong$home_cluster, genome_wide_Eu_per_group_df_qlong$location_id),]
genome_wide_Eu_per_group_df_qlong$data_class = as.factor(genome_wide_Eu_per_group_df_qlong$home_cluster)

genome_wide_Eu_per_group_df_qlong$pop_colour = unlist(lapply(genome_wide_Eu_per_group_df_qlong$Q_type, Q_to_colour))


print("Possibly split in two based on the NFL and LAB")
genome_wide_Eu_per_group_df_qlong = genome_wide_Eu_per_group_df_qlong[order(genome_wide_Eu_per_group_df_qlong$location_id, genome_wide_Eu_per_group_df_qlong$fish_id),]
genome_wide_lab_dat_Eu_per_group_admix = genome_wide_Eu_per_group_df_qlong[genome_wide_Eu_per_group_df_qlong$home_cluster == "LAB",]
genome_wide_nfl_dat_Eu_per_group_admix = genome_wide_Eu_per_group_df_qlong[genome_wide_Eu_per_group_df_qlong$home_cluster == "NFL",]
genome_wide_nor_dat_Eu_per_group_admix = genome_wide_Eu_per_group_df_qlong[genome_wide_Eu_per_group_df_qlong$home_cluster == "NOR",]

genome_wide_LAB_Eu_per_group_admix = ggplot(data=genome_wide_lab_dat_Eu_per_group_admix, aes(x=reorder(fish_id,  home_cluster), y=Q, fill=home_cluster))+
  geom_bar(stat = "identity", fill = genome_wide_lab_dat_Eu_per_group_admix$pop_colour)+
  scale_y_continuous(expand = c(0,0))+
  ylab("Admixture proportion")+
  xlab(" ")+
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.x = element_blank(),
        strip.text = element_text(colour="black", size=12),
        panel.grid = element_blank(),
        panel.background = element_blank(),
        legend.title = element_blank(),
        legend.text = element_text(size = 12))
ggsave('external/FigS7ai_LAB_FST_Eu_per_group_admix.tiff', genome_wide_LAB_Eu_per_group_admix, height = 4, width = 4)
ggsave('external/FigS7ai_LAB_FST_Eu_per_group_admix.jpeg', genome_wide_LAB_Eu_per_group_admix,  height = 4, width = 4)



genome_wide_NFL_Eu_per_group_admix = ggplot(data=genome_wide_nfl_dat_Eu_per_group_admix, aes(x=reorder(fish_id,  home_cluster), y=Q, fill=home_cluster))+
  geom_bar(stat = "identity", fill = genome_wide_nfl_dat_Eu_per_group_admix$pop_colour)+
  scale_y_continuous(expand = c(0,0))+
  ylab("Admixture proportion")+
  xlab(" ")+
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.x = element_blank(),
        strip.text = element_text(colour="black", size=12),
        panel.grid = element_blank(),
        panel.background = element_blank(),
        legend.title = element_blank(),
        legend.text = element_text(size = 12))
ggsave('external/FigS7aii_NFL_FST_Eu_per_group_admix.tiff', genome_wide_NFL_Eu_per_group_admix, height = 4, width = 12)
ggsave('external/FigS7aii_NFL_FST_Eu_per_group_admix.jpeg', genome_wide_NFL_Eu_per_group_admix,  height = 4, width = 12)



genome_wide_NOR_Eu_per_group_admix = ggplot(data=genome_wide_nor_dat_Eu_per_group_admix, aes(x=reorder(fish_id,  home_cluster), y=Q, fill=home_cluster))+
  geom_bar(stat = "identity", fill = genome_wide_nor_dat_Eu_per_group_admix$pop_colour)+
  scale_y_continuous(expand = c(0,0))+
  ylab("Admixture proportion")+
  xlab(" ")+
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.x = element_blank(),
        strip.text = element_text(colour="black", size=12),
        panel.grid = element_blank(),
        panel.background = element_blank(),
        legend.title = element_blank(),
        legend.text = element_text(size = 12))
ggsave('external/FigS7aiii_NOR_LAB_FST_admix.tiff', genome_wide_NOR_Eu_per_group_admix, height = 4, width = 8)
ggsave('external/FigS7aiii_NOR_LAB_FST_admix.jpeg', genome_wide_NOR_Eu_per_group_admix,  height = 4, width = 8)








#B)  LAB-NFL-EU - FST
Eu_per_group_Q2
Eu_per_group_df_qlong = gather(EU_NA_df_Q2, Q_type, Q, Q1, Q2) 
Eu_per_group_df_qlong = Eu_per_group_df_qlong[order(Eu_per_group_df_qlong$home_cluster, Eu_per_group_df_qlong$location_id),]
Eu_per_group_df_qlong$data_class = as.factor(Eu_per_group_df_qlong$home_cluster)

Eu_per_group_df_qlong$pop_colour = unlist(lapply(Eu_per_group_df_qlong$Q_type, Q_to_colour))


print("Possibly split in two based on the NFL and LAB")
Eu_per_group_df_qlong = Eu_per_group_df_qlong[order(Eu_per_group_df_qlong$location_id, Eu_per_group_df_qlong$fish_id),]
lab_dat_Eu_per_group_admix = Eu_per_group_df_qlong[Eu_per_group_df_qlong$home_cluster == "LAB",]
nfl_dat_Eu_per_group_admix = Eu_per_group_df_qlong[Eu_per_group_df_qlong$home_cluster == "NFL",]
nor_dat_Eu_per_group_admix = Eu_per_group_df_qlong[Eu_per_group_df_qlong$home_cluster == "NOR",]

LAB_Eu_per_group_admix = ggplot(data=lab_dat_Eu_per_group_admix, aes(x=reorder(fish_id,  home_cluster), y=Q, fill=home_cluster))+
  geom_bar(stat = "identity", fill = lab_dat_Eu_per_group_admix$pop_colour)+
  scale_y_continuous(expand = c(0,0))+
  ylab("Admixture proportion")+
  xlab(" ")+
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.x = element_blank(),
        strip.text = element_text(colour="black", size=12),
        panel.grid = element_blank(),
        panel.background = element_blank(),
        legend.title = element_blank(),
        legend.text = element_text(size = 12))
ggsave('external/FigS7ci_LAB_FST_Eu_per_group_admix.tiff', LAB_Eu_per_group_admix, height = 4, width = 4)
ggsave('external/FigS7ci_LAB_FST_Eu_per_group_admix.jpeg', LAB_Eu_per_group_admix,  height = 4, width = 4)



NFL_Eu_per_group_admix = ggplot(data=nfl_dat_Eu_per_group_admix, aes(x=reorder(fish_id,  home_cluster), y=Q, fill=home_cluster))+
  geom_bar(stat = "identity", fill = nfl_dat_Eu_per_group_admix$pop_colour)+
  scale_y_continuous(expand = c(0,0))+
  ylab("Admixture proportion")+
  xlab(" ")+
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.x = element_blank(),
        strip.text = element_text(colour="black", size=12),
        panel.grid = element_blank(),
        panel.background = element_blank(),
        legend.title = element_blank(),
        legend.text = element_text(size = 12))
ggsave('external/FigS7bii_NFL_FST_Eu_per_group_admix.tiff', NFL_Eu_per_group_admix, height = 4, width = 12)
ggsave('external/FigS7bii_NFL_FST_Eu_per_group_admix.jpeg', NFL_Eu_per_group_admix,  height = 4, width = 12)



NOR_Eu_per_group_admix = ggplot(data=nor_dat_Eu_per_group_admix, aes(x=reorder(fish_id,  home_cluster), y=Q, fill=home_cluster))+
  geom_bar(stat = "identity", fill = nor_dat_Eu_per_group_admix$pop_colour)+
  scale_y_continuous(expand = c(0,0))+
  ylab("Admixture proportion")+
  xlab(" ")+
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.x = element_blank(),
        strip.text = element_text(colour="black", size=12),
        panel.grid = element_blank(),
        panel.background = element_blank(),
        legend.title = element_blank(),
        legend.text = element_text(size = 12))
ggsave('external/FigS7biii_NOR_LAB_FST_admix.tiff', NOR_Eu_per_group_admix, height = 4, width = 8)
ggsave('external/FigS7biii_NOR_LAB_FST_admix.jpeg', NOR_Eu_per_group_admix,  height = 4, width = 8)






#C)  LAB-NFL - FST
#this one doesn't make sense to me, why do the differentiated loci not separate out the populations?


base_NA_df

base_NA_df_qlong = gather(base_NA_df, Q_type, Q, Q1, Q2) 
base_NA_df_qlong = base_NA_df_qlong[order(base_NA_df_qlong$home_cluster, base_NA_df_qlong$location_id),]
base_NA_df_qlong$data_class = as.factor(base_NA_df_qlong$home_cluster)


base_NA_df_qlong$pop_colour = unlist(lapply(base_NA_df_qlong$Q_type, Q_to_colour))

print("Possibly split in two based on the NFL and LAB")
base_NA_df_qlong = base_NA_df_qlong[order(base_NA_df_qlong$location_id, base_NA_df_qlong$fish_id),]
lab_dat_base_NA_admix = base_NA_df_qlong[base_NA_df_qlong$home_cluster == "LAB",]
nfl_dat_base_NA_admix = base_NA_df_qlong[base_NA_df_qlong$home_cluster == "NFL",]

LAB_base_NA_admix = ggplot(data=lab_dat_base_NA_admix, aes(x=reorder(fish_id,  home_cluster), y=Q, fill=home_cluster))+
  geom_bar(stat = "identity", fill = lab_dat_base_NA_admix$pop_colour)+
  scale_y_continuous(expand = c(0,0))+
  ylab("Admixture proportion")+
  xlab(" ")+
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.x = element_blank(),
        strip.text = element_text(colour="black", size=12),
        panel.grid = element_blank(),
        panel.background = element_blank(),
        legend.title = element_blank(),
        legend.text = element_text(size = 12))
ggsave('external/FigS7ci_NFL_LAB_FST_admix.tiff', LAB_base_NA_admix, height = 4, width = 4)
ggsave('external/FigS7ci_NFL_LAB_FST_admix.jpeg', LAB_base_NA_admix,  height = 4, width = 4)

NFL_base_NA_admix = ggplot(data=nfl_dat_base_NA_admix, aes(x=reorder(fish_id,  home_cluster), y=Q, fill=home_cluster))+
  geom_bar(stat = "identity", fill = nfl_dat_base_NA_admix$pop_colour)+
  scale_y_continuous(expand = c(0,0))+
  ylab("Admixture proportion")+
  xlab(" ")+
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.x = element_blank(),
        strip.text = element_text(colour="black", size=12),
        panel.grid = element_blank(),
        panel.background = element_blank(),
        legend.title = element_blank(),
        legend.text = element_text(size = 12))
ggsave('external/FigS7cii_NFL_LAB_FST_admix.tiff', NFL_base_NA_admix, height = 4, width = 12)
ggsave('external/FigS7cii_NFL_LAB_FST_admix.jpeg', NFL_base_NA_admix,  height = 4, width = 12)






library(patchwork)


print("patchwork version")
genome_wide_NFL_Eu_per_group_admix = genome_wide_NFL_Eu_per_group_admix + labs(tag = "A")
NFL_Eu_per_group_admix = NFL_Eu_per_group_admix + labs(tag = "B")
NFL_base_NA_admix = NFL_base_NA_admix + labs(tag = "C")

outplot = ( genome_wide_NFL_Eu_per_group_admix | genome_wide_LAB_Eu_per_group_admix | genome_wide_NOR_Eu_per_group_admix ) /
          ( NFL_Eu_per_group_admix | LAB_Eu_per_group_admix | NOR_Eu_per_group_admix ) /
          ( NFL_base_NA_admix | LAB_base_NA_admix )

ggsave('external/FigS7.tiff', outplot, height = 20, width = 30)
ggsave('external/FigS7.jpeg', outplot,  height = 20, width = 30)


























##############
# Admixture tract -  idea below
# scrapping for now as its very convoluted and relies on a bunch of data manipulation (phasing,  file conversion, etc.)

#####################
# seems like I need to run an admixture proportion analysis for the Eu - NA lineage data
# as I figure these are the steps I need to figure out


#[x] 1. PED -> VCF
# do it for all fish from the admixture analysis, use plink to accomplish this

# cd data/raw/
# 
# plink --bfile wildNA_220k_filtered --chr-set 30 no-xy --recode vcf --out wildNA_220k_filtered
# 
# plink --bfile /mnt/c/Users/camnu/bin/salmon-euro-introgression/data/May2022_SNPchip_220K/SNPchip_100MS_SNPpanel_King7_Compare/SNPchip_220K/wild_Salmo220K_NA_EU_LE_filtered --chr-set 30 no-xy --recode vcf --out wildNA_AND_Eu_220k_filtered



#[] 2. Beagle 4.1 to phase the data
#https://faculty.washington.edu/browning/beagle/b4_1.html
# 
# As per Tony:
# 
# cat Ssal31chroms.tsv | parallel --jobs 4 'java -Xmx400G \
#  -jar /datadisk0/software/beagle.27Jan18.7e1.jar \
#  ref=imputegenos/phasedSalmo.salar.imputeref.allrefimpute.{}.vcf.gz \
#  gl=seaagelc/Salmo.salar.seaage.seaage.impute.{}.vcf.gz \
#  out=imputeseaagegp/{}
# 
#Change locations, inputs, and you'll want to switch to gl to gt
# this should exist, as it was done in Bradbury et al. 2022, had 4993 NA samples in that
# now there are 5441 samples, so about 500 more.
# ^apparently code doesn't get saved... seems repeatable


#[] 3. Run admixture proportion analysis to quantify the presence, position, and size 
#  of European ancestry tracts across the genomes of individuals
# use loter: #https://github.com/bcm-uga/Loter

# seems well documented 


# isolate this to just the LAB and NFL fish, as well as Eu if needed(?)

#a. 
#b. Eu Fish and the LAB
#c. Eu Fish and the NFL
#d. Eu Fish and the MAR
#e. all NA baseline - use this to look at the admixture on a per location basis and Id evidence of directionality.



#[] 4. Load the data and visualize/quantify the distributions
#summary stats : proportion per chr (different maxima in the LAB and NFL?)
# visual: sort by pop and plot the tracts (possibly for specific chromosomes)



