


#########################
# libraries
library(tidyverse)
#install.packages("ggforce")
library(ggforce)
library(ggplot2)
library(rnaturalearth)
library(rnaturalearthdata)


############
# set path
data_path = "C:/Users/camnu/bin/salmon-genomic-vulnerability/data/"
setwd(data_path)

alter_cluster_name = function(x){
  if(x == "St. Lawrence"){
    return("Maritimes")
  }else{
    return(x)
  }
}

alter_cluster_IDs = function(x){
  if(x == "STL"){
    return("Maritimes")
  }else if (x == "NFL"){
    return("Newfoundland")
  }else if (x == "LAB"){
    return("Labrador")
  }else{
    return(x)
  }
}




#
# using admixture sub run to quantify the effect of European ancestry on the NFL/LAB differences
# 
# 1. get list of significant fst outliers between NFL and LAB
#   - these are the regions of the genome that are different
# 

full_fst_file = "interim/fst_files/LAB_v_NFL_fst_full_pairwise_updated_genome.fst"

full_fst = read_tsv(full_fst_file)
mean_fst = mean(full_fst$FST, na.rm = TRUE)
sd_fst = sd(full_fst$FST, na.rm = TRUE)

genome_thresh = mean_fst + (3*sd_fst)

full_fst$high_fst = unlist(lapply(1:nrow(full_fst), function(i){
  full_fst$FST[[i]] > genome_thresh
}))

high_fst = full_fst[full_fst$high_fst==TRUE,]
high_fst #1648 markers with Fst score > 3sd from the mean for the pairwise comparion
print("1648 markers with Fst score > 3sd from the mean for the pairwise comparion")


#read in the full snp list, then subset everything but the above to make an exclude list.
full_snps = read_tsv("C:/Users/camnu/bin/salmon-euro-introgression/data/May2022_SNPchip_220K/SNPchip_100MS_SNPpanel_King7_Compare/SNPchip_220K/all_fish_Salmo220K_Polyhigh_maf01.map", 
                        col_names = c("CHR", "SNP", "MAP", "POS"))

exclude_snps = full_snps[!full_snps$SNP %in% high_fst$SNP, ]


write.table(exclude_snps$SNP, "Fst_LAB_NFL_exclude.tsv", col.names = F, row.names = F, quote = F, sep = "\t")


# 2. Get the list of fish in the NFL and LAB and EU populations
#   - use new and old metadata, these are the fish we will use to analyze admixture
# 


clusters_per_location  = read_tsv("interim/cluster_and_q_data_combined.tsv")
clusters_per_location$Population = unlist(lapply(clusters_per_location$river_majority_cluster, alter_cluster_IDs))
clusters_per_location$location_id = clusters_per_location$SiteCode


nfl_and_lab = clusters_per_location[clusters_per_location$home_cluster != "STL",]

full_admix = read_tsv("C:/Users/camnu/bin/salmon-euro-introgression/data/May2022_SNPchip_220K/SNPchip_100MS_SNPpanel_King7_Compare/SNPchip_220K/all_data_admix_vals.tsv")
table(full_admix$data_class)
nor_fish = full_admix[full_admix$data_class == "Norway",]


print("write the combined list to a file")

out_df = rbind(nfl_and_lab[c("location_id", "fish_id")], nor_fish[c("location_id", "fish_id")])

names(out_df) = c("FID", "IID")
write_tsv(out_df, "Nor_NFL_LAB_individual_list.tsv")

base_out = nfl_and_lab[c("location_id", "fish_id")]
names(base_out) = c("FID", "IID")

write_tsv(base_out, "NFL_LAB_individual_list.tsv")

print("Admixture with:")
print("806 Norwegian")
print("431 LAB")
print("3244 NFL")

# 
# 3. use plink to subset the old bed/bim file (complete data set from SalmonEuAdmix)
#   - pull just the Fst outliers and the LAB/NFL/Eu fish
#   - pull just the Fst outleris and the LAB/NFL fish
# 


#plink --bfile /mnt/c/Users/camnu/bin/salmon-euro-introgression/data/May2022_SNPchip_220K/SNPchip_100MS_SNPpanel_King7_Compare/SNPchip_220K/wild_Salmo220K_NA_EU_LE_filtered --keep Nor_NFL_LAB_individual_list.tsv --exclude Fst_LAB_NFL_exclude.tsv --chr-set 30 no-xy --make-bed --out Nor_NFL_LAB_220K_filtered
## worked: 1634 variants and 4481 samples pass filters and QC.


#plink --bfile /mnt/c/Users/camnu/bin/salmon-euro-introgression/data/May2022_SNPchip_220K/SNPchip_100MS_SNPpanel_King7_Compare/SNPchip_220K/wild_Salmo220K_NA_EU_LE_filtered --keep NFL_LAB_individual_list.tsv --exclude Fst_LAB_NFL_exclude.tsv --chr-set 30 no-xy --make-bed --out NFL_LAB_220K_filtered
## worked: 1634 variants and 3675 samples pass filters and QC.



# 4. run admixture
# 
# - k=2 and k=3 -> Fst outliers and the LAB/NFL/Eu fish ()
#     - if the admixture proprotions are really high, then can say the regions of European contact are driving the differentiation between the groups
#     - if admixture proportions are low, can say that the Eu admixture is not in the differentiated parts of the genome, likely shared Eu ancestry.
# - k=2         ->  Fst outleris and the LAB/NFL fish (serves as a baseline)



# admixture Nor_NFL_LAB_220K_filtered.bed 2
# admixture Nor_NFL_LAB_220K_filtered.bed 3
# 
# admixture NFL_LAB_220K_filtered.bed 2



#5. load the admixture outputs, plot and analyze the results












##############
# Tony's idea below
# scrapping for now as its very convoluted and relies on a bunch of data manipulation (phasing,  file conversion, etc.)

#####################
# seems like I need to run an admixture proportion analysis for the Eu - NA lineage data
# as I figure these are the steps I need to figure out


#[x] 1. PED -> VCF
# do it for all fish from the admixture analysis, use plink to accomplish this

# cd data/raw/
# 
# plink --bfile wildNA_220k_filtered --chr-set 30 no-xy --recode vcf --out wildNA_220k_filtered
# 
# plink --bfile /mnt/c/Users/camnu/bin/salmon-euro-introgression/data/May2022_SNPchip_220K/SNPchip_100MS_SNPpanel_King7_Compare/SNPchip_220K/wild_Salmo220K_NA_EU_LE_filtered --chr-set 30 no-xy --recode vcf --out wildNA_AND_Eu_220k_filtered



#[] 2. Beagle 4.1 to phase the data
#https://faculty.washington.edu/browning/beagle/b4_1.html
# 
# As per Tony:
# 
# cat Ssal31chroms.tsv | parallel --jobs 4 'java -Xmx400G \
#  -jar /datadisk0/software/beagle.27Jan18.7e1.jar \
#  ref=imputegenos/phasedSalmo.salar.imputeref.allrefimpute.{}.vcf.gz \
#  gl=seaagelc/Salmo.salar.seaage.seaage.impute.{}.vcf.gz \
#  out=imputeseaagegp/{}
# 
#Change locations, inputs, and you'll want to switch to gl to gt
# this should exist, as it was done in Bradbury et al. 2022, had 4993 NA samples in that
# now there are 5441 samples, so about 500 more.
# ^apparently code doesn't get saved... seems repeatable


#[] 3. Run admixture proportion analysis to quantify the presence, position, and size 
#  of European ancestry tracts across the genomes of individuals
# use loter: #https://github.com/bcm-uga/Loter

# seems well documented 


# isolate this to just the LAB and NFL fish, as well as Eu if needed(?)

#a. 
#b. Eu Fish and the LAB
#c. Eu Fish and the NFL
#d. Eu Fish and the MAR
#e. all NA baseline - use this to look at the admixture on a per location basis and Id evidence of directionality.



#[] 4. Load the data and visualize/quantify the distributions
#summary stats : proportion per chr (different maxima in the LAB and NFL?)
# visual: sort by pop and plot the tracts (possibly for specific chromosomes)



