setwd("C:/Users/camnu/bin/salmon-genomic-vulnerability/data/")

library('tidyverse')


readPedMap_tsv_fmt = function(ped_file, map_file = "", headers = FALSE){
  if(headers == TRUE){
    #
    map_data = read_tsv(map_file, skip = 1, col_names = c("chromosome", "snp" , "genetic_distance", "physical_distance"))
    
    header_data = c('#family', 'individual', 'sire', 'dam', 'sex', 'pheno')
    header_data = c(header_data, map_data$snp)
    
    snp_data = read_tsv(ped_file, skip = 1, col_names = header_data)
    return(snp_data)
    
  }else{
    
    map_data = read_tsv(map_file, col_names = c("chromosome", "snp" , "genetic_distance", "physical_distance"))
    
    header_data = c('#family', 'individual', 'sire', 'dam', 'sex', 'pheno')
    header_data = c(header_data, map_data$snp)
    
    snp_data = read_tsv(ped_file, col_names = header_data)
    return(snp_data)
  }
  
}



meta_data = read_tsv('raw/Salmo_220K_Merged2022_ENVmatchyear_Metadata.txt')      

dim(meta_data)

# make ped and map file in case there are needed
#plink --bfile Salmo_220K_Merged2022_COMPLETE_ENVmatchyear --recode --tab --out Salmo_220K_Merged2022_COMPLETE_ENVmatchyear --chr-set 30 no-xy


# read in the raw file and have a look at what is in that
# appears to be dosage encoded genotypes with the main PED leading columns
geno_data = read_table('raw/Salmo_220K_Merged2022_COMPLETE_ENVmatchyear.raw')      

dim(geno_data) #3865 individuals

names(geno_data)[1:10] #leading ped cols, then dosage GTs

geno_data[1:5, 1:10] #dosage encoded

geno_data$FID[1:10] #family IDs

geno_data$IID[1:10] #individual IDs - this is what the data should be joined to the meta data with


print("check the matching of the geno and meta ids")

length(geno_data$IID) #3865

length(meta_data$ID) #5570

matches = sum(geno_data$IID %in% meta_data$ID)

matches == length(geno_data$IID)
#all geno_individuals have matching metadata





