#Cam's notes on filtering the bed file

# go to the location with the files
#cd /directory_with_the_files/


# trying to turn the bed file into a ped so you can look at it
# plink --bfile Salmo_220K_2022_RivAge_24 --recode tab --out Salmo_220K_2022_RivAge_24 --chr-set 30 no-xy
# Above fails because of those weird SNPs
## Error: Invalid chromosome code 'NA' on line 186293 of .bim file.
## (Use --allow-extra-chr to force it to be accepted.)

## 1. I added a line to drop the five trailing SNPs with no chr that were causing issues
# plink --bfile Salmo_220K_2022_RivAge_24 --chr-set 30 no-xy --make-bed --maf 0.01 --out maf_filtered_Salmo_220K_2022_RivAge_24 --allow-extra-chr --exclude exclude_snps.txt  

## 2. Make a ped, in case you want to look at the data that pass QC
# plink --bfile maf_filtered_Salmo_220K_2022_RivAge_24 --recode tab --out maf_filtered_Salmo_220K_2022_RivAge_24 --chr-set 30 no-xy

## works:
## Total genotyping rate is 0.998082.
## 152832 variants and 925 samples pass filters and QC.

## 3. Optional: -filter off linkage disequlibrim
## NB I run --bfile not --file and then it uses the bed as input and works a little faster
## a. make the prune file
#plink --bfile maf_filtered_Salmo_220K_2022_RivAge_24 --indep-pairwise 50 5 0.5 --out maf_filtered_Salmo_220K_2022_RivAge_24_window_LD --chr-set 30 no-xy --allow-extra-chr
## b. prune the SNPS
#plink --bfile maf_filtered_Salmo_220K_2022_RivAge_24 --exclude maf_filtered_Salmo_220K_2022_RivAge_24_window_LD.prune.out --out maf_filtered_Salmo_220K_2022_RivAge_24_LD_filtered --make-bed --chr-set 30 no-xy --allow-extra-chr


# This will make a file set named maf_filtered_Salmo_220K_2022_RivAge_24_LD_filtered.*
# you could re-run the ped command on it to turn it into a ped file.


#4. Okay I've got the filtered data now what?

## you might want to do some filtering based off of individuals (i.e. different subsets by location or something)
## to do this, you need a file with the individual IDs you want to REMOVE
## each ID on a separate line, here in an imaginary file named: exclude_these_fish.tsv
# plink --bfile maf_filtered_Salmo_220K_2022_RivAge_24_LD_filtered --exclude exclude_these_fish.tsv --out data_but_some_fish_removed --make-bed --chr-set 30 no-xy


# you can run a PCA on it in R using pcadapt
install.packages('pcadapt')
library(pcadapt)
# for more info: https://bcm-uga.github.io/pcadapt/articles/pcadapt.html

# run the PCA on the filtered SNP file
demo_salmon = read.pcadapt("maf_filtered_Salmo_220K_2022_RivAge_24_LD_filtered.bed", type = "bed")
demo_salmon_PCA = pcadapt(demo_salmon, K = 2, min.maf = 0.01)
(demo_salmon_PCA$singular.values)^2*100 #these are your PC axes scores

#make a data frame
demo_salmon_PCA_Scores = data.frame(demo_salmon_PCA$scores)
colnames(demo_salmon_PCA_Scores) = c("PC1", "PC2")
print("from here down code will fail as is")
print("here I'm pretending to attach the PC scores to a dataframe of individuals with their metadata")
print("for you this could be locations and run timining and such.")
PC_with_Meta = bind_cols(Meta_data_file, demo_salmon_PCA_Scores) 


dim(PC_with_Meta)

#Figure2B
pca_of_filtered_demo = ggplot() + 
                              geom_point(data = PC_with_Meta, aes(x =PC1, y = PC2, colour = data_class))  +
                                labs(title = "PCA of filtered genetic variation baseline with all individuals")+
                                xlab("PC1 (XX.X% of variance)") +
                                ylab("PC2 (XX.X% variance explained)")+
                                labs(color='Sample Group')+ 
                                
                                theme_classic()
ggsave('demo_pca_plot.png', pca_of_filtered_demo)
