data_path = "C:/Users/camnu/bin/salmon-genomic-vulnerability/data/"
setwd(data_path)


#########################
# libraries

library(tidyverse)
#install.packages("ggforce")
library(ggforce)
library(ggplot2)
theme_set(theme_bw())

library(rnaturalearth)
library(rnaturalearthdata)

#########################
# functions

get_long_fmt_admix = function(meta_k3_Q_vals){
  meta_k3_Q_vals = meta_k3_Q_vals[!is.na(meta_k3_Q_vals$Province), ]
  
  
  table(clustered_full_data_PCA$Province)
  
  k3_qlong = gather(meta_k3_Q_vals, Q_type, Q, Q1, Q2, Q3) 
  k3_qlong =k3_qlong[order(k3_qlong$data_class, k3_qlong$location_id),]
  k3_qlong$data_class = as.factor(k3_qlong$Province)
  
  
  k3_qlong = k3_qlong[order(k3_qlong$Cluster, k3_qlong$Province, k3_qlong$location_id),]
  k3_qlong = k3_qlong[!is.na(k3_qlong$Province), ]
  k3_qlong = k3_qlong[!is.na(k3_qlong$Cluster), ]
  
  return(k3_qlong)  
}

alter_cluster_name = function(x){
  if(x == "St. Lawrence"){
    return("Maritimes")
  }else{
    return(x)
  }
}

alter_cluster_IDs = function(x){
  if(x == "STL"){
    return("Maritimes")
  }else if (x == "NFL"){
    return("Newfoundland")
  }else if (x == "LAB"){
    return("Labrador")
  }else{
    return(x)
  }
}

#########################
# load the data
data_path = "C:/Users/camnu/bin/salmon-genomic-vulnerability/data/"
setwd(data_path)


#################################################################################
#################################################################################
#################################################################################
# figure refinement for paper
#################################################################################
#################################################################################
#################################################################################


#########################
# Figure 1


#####
# Fig 1A - population assignments for each individual

print("Colour scheme and apply to all plots of the clusters")
# salmon a cute touch, but it gets washed out a bit
#colour_coord =        c("#3D5B59",  "#FCB5AC",   "#3EB489" )
#TOo light
#colour_coord =        c("#74BDCB",  "#FFA384",   "#EFE7BC" )

colour_coord =        c("#16ABEE",  "#CC9753",   "#3EB489" )
names(colour_coord) = c("Labrador", "Maritimes", "Newfoundland")


clustered_full_data_PCA = read_tsv("interim/full_data_PCA_kmeans_cluster_info.tsv")


publication_plot_kmeans = function(df , colour_coord){
  
  world = ne_countries(scale = "medium", returnclass = "sf")
  class(world)
  #plot a sub region
  lat_min = min(df$lat, na.rm = TRUE)
  lat_max = max(df$lat, na.rm = TRUE)
  long_min = min(df$lon, na.rm = TRUE)
  long_max = max(df$lon, na.rm = TRUE)
  
  clusters_by_study_range = ggplot(data = world) +
    geom_sf(fill = "antiquewhite1")+
    coord_sf(xlim = c(long_min - 2.5, long_max + 2.5), 
             ylim = c(lat_min - 1, lat_max + 1), expand = FALSE) +
    geom_jitter(data = df, aes(x = lon, y = lat, 
                               colour=as.factor(Population)), 
                position=position_jitter(width = .12, height = 0.12), alpha = 0.5)+
    xlab("Longitude") + ylab("Latitude") +
    theme(legend.position="none",
          #panel.grid.major = element_blank(), 
          #panel.grid.minor = element_blank(), 
          panel.background = element_rect(fill = "#E7F2F8",
                                          colour = "#E7F2F8",
                                          size = 0.5, linetype = "solid"))+
    scale_color_manual(values = colour_coord)
  
  
  return(clusters_by_study_range)
}




clustered_full_data_PCA$Population = unlist(lapply(clustered_full_data_PCA$Cluster, alter_cluster_name))

Fig1A = publication_plot_kmeans(clustered_full_data_PCA, colour_coord)

ggsave('external/Fig1A.tiff', Fig1A, width = 10, height = 8)
ggsave('external/Fig1A.jpeg', Fig1A, width = 10, height = 8)


print("TODO - map with fancier background and no gridlines")
print("PCA with no gridlines, change name to MAR, classification from cluster -> population")


clustered_full_data_PCA

print("final version of figures")


#####
# Alt Fig 1A - majority population assignments for each river
#
clusters_per_location  = read_tsv("interim/cluster_and_q_data_combined.tsv")

clusters_per_location$Population = unlist(lapply(clusters_per_location$river_majority_cluster, alter_cluster_IDs))

dedup_locations = distinct(clusters_per_location[c("SiteCode", "lat", "lon", "Population")])
dedup_locations
dedup_locations = dedup_locations[!duplicated(dedup_locations$SiteCode),]

plot_river_major = function(df, colour_coord){
  
  world = ne_countries(scale = "medium", returnclass = "sf")
  class(world)
  #plot a sub region
  lat_min = min(df$lat, na.rm = TRUE)
  lat_max = max(df$lat, na.rm = TRUE)
  long_min = min(df$lon, na.rm = TRUE)
  long_max = max(df$lon, na.rm = TRUE)
  
  river_major_k = ggplot(data = world) +
    geom_sf(fill = "antiquewhite1")+
    coord_sf(xlim = c(long_min - 2.5, long_max + 2.5), 
             ylim = c(lat_min - 1, lat_max + 1), expand = FALSE) +
    geom_jitter(data = df, aes(x = lon, y = lat, 
                               colour=as.factor(Population)), 
                position=position_jitter(width = .12, height = 0.12))+
    xlab("Longitude") + ylab("Latitude") +
    theme(legend.position="none",
          panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(), 
          panel.background = element_rect(fill = "#E7F2F8",
                                          colour = "#E7F2F8",
                                          size = 0.5, linetype = "solid"))+
    scale_color_manual(values = colour_coord)
  
  
  return(clusters_by_study_range)
}


Alt_Fig1A = plot_river_major(dedup_locations, colour_coord)

ggsave('external/Alt_Fig1A.tiff', Alt_Fig1A, width = 10, height = 8)
ggsave('external/Alt_Fig1A.jpeg', Alt_Fig1A, width = 10, height = 8)

#####
# Fig 1B - PCA of the individuals, with the k-means projections and labels
Fig1B = ggplot() + geom_point(data = clustered_full_data_PCA, aes(x = PC1, y = PC2, colour=Population), alpha = 0.7)+
  geom_mark_ellipse(data=clustered_full_data_PCA, aes(x = PC1, y = PC2, colour = Population, label = Population),
                    expand = unit(0.2,"mm"))+
  xlab("PC1, 3.39% variance explained") +
  ylab("PC2, 1.46% variance explained") +
  scale_color_manual(values = colour_coord)+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())


ggsave('external/Fig1B.tiff', Fig1B, width = 10, height = 8)
ggsave('external/Fig1B.jpeg', Fig1B, width = 10, height = 8)


#####
# Fig 1C - Admixture for three populations side-by-side
meta_k3_Q_vals  = read_tsv("interim/cluster_and_q_data_combined.tsv")


cluster_admix_long = get_long_fmt_admix(meta_k3_Q_vals)

Q_to_name = function(Q){  
  col_dict = c("LAB", "NFL", "MAR")  
  names(col_dict)  =  c("Q3", "Q2", "Q1")
  return(col_dict[[Q]])
}


Q_to_colour = function(Q){  
  col_dict = c("#16ABEE", "#3EB489", "#CC9753")  
  names(col_dict)  =  c("Q3", "Q2", "Q1")
  return(col_dict[[Q]])
}

cluster_admix_long$Population = unlist(lapply(cluster_admix_long$Q_type, Q_to_name))
cluster_admix_long$pop_colour = unlist(lapply(cluster_admix_long$Q_type, Q_to_colour))

mar_dat = cluster_admix_long[cluster_admix_long$river_majority_cluster == "STL",]
nfl_dat = cluster_admix_long[cluster_admix_long$river_majority_cluster == "NFL",]
lab_dat = cluster_admix_long[cluster_admix_long$river_majority_cluster == "LAB",]

table(lab_dat$river_majority_cluster)

mar_admix_bar = ggplot(data=mar_dat, aes(x=reorder(fish_id,  SiteCode), y=Q, fill=Population))+
  geom_bar(stat = "identity", colour = mar_dat$pop_colour)+
  scale_y_continuous(expand = c(0,0))+
  ylab("Admixture proportion")+
  xlab(" ")+
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.x = element_blank(),
        strip.text = element_text(colour="black", size=12),
        panel.grid = element_blank(),
        panel.background = element_blank(),
        legend.position="none",
        legend.title = element_blank(),
        legend.text = element_text(size = 12))

nfl_admix_bar = ggplot(data=nfl_dat, aes(x=reorder(fish_id,  SiteCode), y=Q, fill=Population))+
  geom_bar(stat = "identity", colour = nfl_dat$pop_colour)+
  scale_y_continuous(expand = c(0,0))+
  ylab("")+
  xlab(" ")+
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.x = element_blank(),
        strip.text = element_text(colour="black", size=12),
        panel.grid = element_blank(),
        panel.background = element_blank(),
        legend.position="none",
        legend.title = element_blank(),
        legend.text = element_text(size = 12))


lab_dat = lab_dat[order(lab_dat$SiteCode, lab_dat$fish_id),]

lab_dat$pop_colour[lab_dat$Population == "LAB"]

lab_admix_bar = ggplot(data=lab_dat, aes(x=reorder(fish_id, SiteCode), y=Q, fill=Population))+
  geom_bar(stat = "identity", colour = lab_dat$pop_colour)+
  scale_y_continuous(expand = c(0,0))+
  ylab("")+
  xlab(" ")+
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.x = element_blank(),
        strip.text = element_text(colour="black", size=12),
        panel.grid = element_blank(),
        panel.background = element_blank(),
        legend.position="none",
        legend.title = element_blank(),
        legend.text = element_text(size = 12))

#scale widths proportionally to their number of samples
(nrow(mar_dat) / nrow(cluster_admix_long)) *20
(nrow(nfl_dat) / nrow(cluster_admix_long)) *20
(nrow(lab_dat) / nrow(cluster_admix_long)) *20

ggsave('external/Fig1C_i.tiff', mar_admix_bar, height = 4, width = 6.5)
ggsave('external/Fig1C_i.jpeg', mar_admix_bar,  height = 4, width = 6.5)

ggsave('external/Fig1C_ii.tiff', nfl_admix_bar,  height = 4, width = 12)
ggsave('external/Fig1C_ii.jpeg', nfl_admix_bar,  height = 4, width = 12)

ggsave('external/Fig1C_iii.tiff', lab_admix_bar,  height = 4, width = 1.5)
ggsave('external/Fig1C_iii.jpeg', lab_admix_bar,  height = 4, width = 1.5)



#########################
# Figure 2 - 
#install.packages("scatterpie")
library(scatterpie)

#####
# Fig 2A - Map of the pie charts of per location combined genetic admxiture attribution
meta_k3_Q_vals  = read_tsv("interim/cluster_and_q_data_combined.tsv")

col_dict = c("LAB", "NFL", "MAR")  
names(col_dict)  =  c("Q3", "Q2", "Q1")


Q_proportions = meta_k3_Q_vals %>% 
                  group_by(SiteCode) %>%
                    summarize(Labrador = sum(Q3) / n(), 
                              Newfoundland = sum(Q2) / n(), 
                              Maritimes = sum(Q1) / n(),
                              count = n())            

pie_map = inner_join(distinct(Q_proportions), distinct(meta_k3_Q_vals[c("SiteCode", "lat", "lon")]))
pie_map = pie_map[!duplicated(pie_map$SiteCode),]

write_tsv(pie_map, "interim/df_for_pie_charts.tsv")

print("having trouble combining the two components, bodge below")
#df = pie_map
admixture_plot = function(df){
  world = ne_countries(scale = "medium", returnclass = "sf")
  
  class(world)
  #plot a sub region
  lat_min = min(df$lat, na.rm = TRUE)
  lat_max = max(df$lat, na.rm = TRUE)
  long_min = min(df$lon, na.rm = TRUE)
  long_max = max(df$lon, na.rm = TRUE)
  
  study_range = ggplot(data = world)  +    
                              geom_sf(fill = "antiquewhite1")+
    coord_sf(xlim = c(long_min - 2.5, long_max + 2.5), 
             ylim = c(lat_min - 1, lat_max + 1), expand = FALSE) +
      xlab("Longitude") + ylab("Latitude") +
    theme(legend.position="none",
          #panel.grid.major = element_blank(), 
          #panel.grid.minor = element_blank(), 
          panel.background = element_rect(fill = "#E7F2F8",
                                          colour = "#E7F2F8",
                                          size = 0.5, linetype = "solid"))

  ggsave('external/Fig2A_back.tiff', study_range, width = 10, height = 8)
  ggsave('external/Fig2A_back.jpeg', study_range,width = 10, height = 8)
  
  clusters_and_pie = ggplot(data = world)  +  
                      geom_scatterpie(data=df, aes(x=lon, y=lat, group=SiteCode, r=0.3),
                                      cols= c("Labrador", "Newfoundland", "Maritimes")) + coord_equal() + 
                      scale_fill_manual(values = colour_coord) +
                      xlim(long_min- 2.5, long_max+ 2.5)+ ylim(lat_min-1, lat_max+1)+
                      theme(legend.position="none",
                            panel.grid.major = element_blank(), 
                            panel.grid.minor = element_blank())    +
                      xlab("Longitude") + ylab("Latitude")   
  ggsave('external/Fig2A_front.tiff', clusters_and_pie, width = 10, height = 8)
  ggsave('external/Fig2A_front.jpeg', clusters_and_pie, width = 10, height = 8)
  
}

map_clusters_and_pie = ggplot(data = world) + 
                        geom_sf(fill = "antiquewhite1")+
  coord_sf(xlim = c(long_min - 2.5, long_max + 2.5), 
           ylim = c(lat_min - 1, lat_max + 1), expand = FALSE)+  
  geom_scatterpie(data=df, aes(x=lon, y=lat, group=SiteCode, r=0.3),
                  cols= c("Labrador", "Newfoundland", "Maritimes")) + 
  scale_fill_manual(values = colour_coord) +
  xlim(long_min- 2.5, long_max+ 2.5)+ ylim(lat_min-1, lat_max+1)+
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_rect(fill = "#E7F2F8",
                                        colour = "#E7F2F8",
                                        size = 0.5, linetype = "solid"))+
  xlab("Longitude") + ylab("Latitude")   

ggsave('external/Fig2A_both.tiff', map_clusters_and_pie, width = 16, height = 10)



#######################################
#2B - map of admixture hotspots and the bathymetric data / glaciers
world = ne_countries(scale = "medium", returnclass = "sf")

class(world)
#plot a sub region
lat_min = min(df$lat, na.rm = TRUE)
lat_max = max(df$lat, na.rm = TRUE)
long_min = min(df$lon, na.rm = TRUE)
long_max = max(df$lon, na.rm = TRUE)

#' df = meta_k3_Q_vals

plot_high_admixture = function(df){
  high_admix_individuals = ggplot(data = world)  +    
    geom_sf(fill = "antiquewhite1")+
    coord_sf(xlim = c(long_min - 2.5, long_max + 2.5), 
             ylim = c(lat_min - 1, lat_max + 1), expand = FALSE) +
    xlab("Longitude") + ylab("Latitude") +
    theme(#panel.grid.major = element_blank(), 
          #panel.grid.minor = element_blank(), 
          panel.background = element_rect(fill = "#E7F2F8",
                                          colour = "#E7F2F8",
                                          size = 0.5, linetype = "solid")) +
    geom_jitter(data = df, aes(x = lon, y = lat, 
                               colour=as.factor(low_home_q)), 
                position=position_jitter(width = .12, height = 0.12), alpha = 0.3)+
    labs(color = "High Admixture")
  
  return(high_admix_individuals)
}

Fig2B = plot_high_admixture(meta_k3_Q_vals)
ggsave('external/Fig2B.tiff', Fig2B, width = 10, height = 8)
ggsave('external/Fig2B.jpeg', Fig2B, width = 10, height = 8)


print("TODO - make a bathymetry version of the sample locations for use in supplementary or somewhere else")

print("take this one and do the admixture plot, show the locations relative to the bathyetry")
print("also remake the sample location and number plot with this one")

#devtools::install_github("MikkoVihtakari/ggOceanMapsData") 
library(ggOceanMapsData)
library(ggOceanMaps)
library(readxl)
library(tidyverse)
library(ggspatial)

colour_coord =        c("#16ABEE",  "#CC9753",   "#3EB489" )
names(colour_coord) = c("Labrador", "Maritimes", "Newfoundland")

bathy_publication_plot_kmeans = function(df , colour_coord){
  
  world = ne_countries(scale = "medium", returnclass = "sf")
  class(world)
  #plot a sub region
  lat_min = min(df$lat, na.rm = TRUE)
  lat_max = max(df$lat, na.rm = TRUE)
  long_min = min(df$lon, na.rm = TRUE)
  long_max = max(df$lon, na.rm = TRUE)
  

  bath_variant = ggOceanMaps::basemap(limits = c(long_min - 1.5,long_max + 1.5, lat_min - 1, lat_max + 1 ), 
                                      land.col = "antiquewhite1", bathymetry = TRUE, rotate = TRUE)+
                      ggspatial::geom_spatial_point(data = df,  aes(x = lon,  y = lat, color = Population), 
                                            size = 3, alpha = 0.5, position=position_jitter(width = .25, height = 0.25))+
                      scale_color_manual(values = colour_coord)
  
  return(bath_variant)
}


plot_admix_smmary = function(df, main_title = "", 
                             subtitle = ""){
  
  
  world = ne_countries(scale = "medium", returnclass = "sf")
  class(world)
  #plot a sub region
  lat_min = min(df$lat, na.rm = TRUE)
  lat_max = max(df$lat, na.rm = TRUE)
  long_min = min(df$lon, na.rm = TRUE)
  long_max = max(df$lon, na.rm = TRUE)
  
  clusters_by_study_range = ggplot(data = world) +
    geom_sf() +
    coord_sf(xlim = c(long_min - 2, long_max + 2), 
             ylim = c(lat_min - 2, lat_max + 2), expand = FALSE) +
    geom_jitter(data = df, aes(x = lon, y = lat, 
                               colour=as.factor(low_home_q)), 
                position=position_jitter(width = .25, height = 0.25), alpha = 0.5)+
    labs(title = main_title, 
         subtitle = subtitle, 
         color = "Is High Admixture Individual")
  
  return(clusters_by_study_range)
}



#########################
# Figure 3 - NE and tree

#####
# Fig 3A - Treemix plot, see script 19
print("plot the pie charts next to the names in a grid, can pull from that to place them on the treemix plot")

colour_coord =        c("#16ABEE",  "#CC9753",   "#3EB489" )
names(colour_coord) = c("Labrador", "Maritimes", "Newfoundland")

pie_map = read_tsv("interim/df_for_pie_charts.tsv")

nrow(pie_map)

positions = expand.grid(coordx = seq(from = 1, by = 10, l = 10),
                        coordy = seq(from = 1, by = 10, l = 15))


pie_map$coordx = positions$coordx[1:148]
pie_map$coordy = positions$coordy[1:148]

clusters_and_pie = ggplot()  +  
  geom_scatterpie(data=pie_map, aes(x=coordx, y=coordy, group=SiteCode, r=1),
                  cols= c("Labrador", "Newfoundland", "Maritimes")) + coord_equal() + 
  geom_text(data=pie_map, aes(x=coordx, y=coordy, label=SiteCode), hjust = 1.5, vjust = .5, size = 2)+
  scale_fill_manual(values = colour_coord) +
  theme(legend.position="none",
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank())


ggsave('external/Fig3A_labels.tiff', clusters_and_pie)
ggsave('external/Fig3A_labels.jpeg', clusters_and_pie)



#####
# Fig 3B - GoNe past plot - see script 22



#########################
# Figure 4 - highlighting key regions
library(ggman)
#####
# Fig 4A - P values of SNPs corresponding to figure 1
loading_full_info = read_tsv("interim/full_samlon_pcadapt_loadings_df.tsv")

loading_full_info$bonf_P = p.adjust(loading_full_info$p_values, method= "bonferroni")

Fig4A = ggman(loading_full_info, snp = "snp", 
                                     bp = "physical_distance", chrom = "chromosome", 
                                     pvalue = "p_values", logTransform = TRUE, 
                                    lineColour = "black", title = "")+
              scale_color_manual(values=c("#16ABEE","#CC9753"))+
              xlab("Chromosome") +
              ylab("-log10(p value))")+
              theme(panel.grid.major = element_blank(), 
                    panel.grid.minor = element_blank())+
              theme_classic()


Fig4Abonf = ggman(loading_full_info, snp = "snp", 
                                    bp = "physical_distance", chrom = "chromosome", 
                                    pvalue = "bonf_P", logTransform = TRUE, 
                                    lineColour = "black", title = "")+
              scale_color_manual(values=c("#16ABEE","#CC9753"))+
              xlab("Chromosome") +
              ylab("-log10(Bonferroni corrected p value)")+
              theme(panel.grid.major = element_blank(), 
                    panel.grid.minor = element_blank())+
              theme_classic()

#omitted, need a value for the threshold?
#geom_hline(yintercept=0.15, linetype="dashed", colour="red", 1) + 

ggsave('external/Fig4A.jpeg', Fig4A,  width = 16, height = 6)
ggsave('external/Fig4A.tiff', Fig4A,  width = 16, height = 6)

ggsave('external/Fig4A_bonf.jpeg', Fig4Abonf,  width = 16, height = 6)
ggsave('external/Fig4A_bonf.tiff', Fig4Abonf,  width = 16, height = 6)

print("the non- corrected looks better, so will use that in the final figure")

#####
# Fig 4B - Pairwise Fst and such
