####
# set the location
####

setwd("C:/Users/camnu/bin/salmon-euro-introgression/scripts/")

data_path = "../data/May2022_SNPchip_220K/SNPchip_100MS_SNPpanel_King7_Compare/SNPchip_220K/"

setwd(data_path)

####
# load necessary libraries
####

library(tidyverse)
library(data.table)
library(pcadapt)

####
# Necessary functions 
####
read_220_snp_meta = function(snp220_meta_file){
  columns = c("location_id", "fish_id", "location_name", "data_class", 
              "lat", "lon", "date_of_genotyping")
  
  data = read_tsv(snp220_meta_file, col_names = columns)
  
}



#########
# run 3 - just the wild NA
# Quantification of population structure and introgression in baseline wild populations in NA
# for the genomic vulnerability work.



###############################
# 1.a. Read in the metadata file
snp220_meta_file  = "CIGENE_220K_Metadata_All_May18_2022.txt"

snp220k_meta_data = read_220_snp_meta(snp220_meta_file)

snp220k_meta_data


###############################
# 1.b. subset the metadata file list to get only the wild individuals from the NA and EU classifications
# omits all of the individuals from aquaculture facilities
NA_only = snp220k_meta_data %>% filter(data_class %in% c("NorthAm"))
NA_only_filter = NA_only %>% select(location_id,  fish_id)
#5570 individuals in this set, matches the graphing work!

###############################
# 1.c. write with the headers dropped (for plink)
outfile = "only_NA_wild_individuals.tsv"
write.table(NA_only_filter, outfile, col.names = F, row.names = F, quote = F, sep = "\t")




###########################
## 1.h use plink to pull just the NA individuals from the filtered high Q input file 
#      this is because the point here is making the baseline for comparison, so we want to keep them as pure of origin as possible

infile = "wild_Salmo220K_NA_EU_filtered_noadmixQ5"
## run this in the command line
#plink --bfile wild_Salmo220K_NA_EU_filtered_noadmixQ5 --chr-set 30 no-xy --keep only_NA_wild_individuals.tsv --make-bed --maf 0.01 --out wildNA_220k_filtered

# 65576 variants and 5455 samples pass filters and QC.



###########################
# 2. PCA of the NA wild individual set 

# read in the fam file
NA_baseline_FAM <- fread("wildNA_220k_filtered.fam") %>% select(V1, V2)
colnames(NA_baseline_FAM) <- c("SiteCode", "fish_id")
NA_baseline_Meta <- inner_join(NA_baseline_FAM, snp220k_meta_data)


write_tsv(NA_baseline_Meta, 'filtered_wild_individuals_meta_data.tsv')
#NA_baseline_Meta = read_tsv('filtered_wild_individuals_meta_data.tsv')

#
# run the PCA on the filtered SNP file
input_NA_pca = read.pcadapt("wildNA_220k_filtered.bed", type = "bed")
NA_wild_PCA = pcadapt(input_NA_pca, K = 2, min.maf = 0.01)
(NA_wild_PCA$singular.values)^2*100 # 28.641487  1.926018


NA_wild_PCA_Scores = data.frame(NA_wild_PCA$scores)
colnames(NA_wild_PCA_Scores) = c("PC1", "PC2")
NA_PC_Meta = bind_cols(NA_baseline_Meta, NA_wild_PCA_Scores) 

dim(NA_PC_Meta)

write_tsv(NA_PC_Meta, "full_wild_PCA_and_meta.tsv")

#Figure2B
pca_of_filtered_NA_wild = ggplot() + 
                              geom_point(data = NA_PC_Meta, aes(x =PC1, y = PC2, colour = data_class))  +
                                labs(title = "PCA of filtered genetic variation - North American wild individuals",
                                     subtitle = "5455 individuals, 65576 SNPs")+
                                xlab("PC1 (3.39% of variance)") +
                                ylab("PC2 (1.46% variance explained)")+
                                labs(color='Sample Group')+ 
                                theme_classic()
ggsave('../../../external/wild_NA_only_pca_nocolour.png', pca_of_filtered_eu_and_NA)


location_pca_of_filtered_NA_wild = ggplot() + 
  geom_point(data = NA_PC_Meta, aes(x =PC1, y = PC2, colour = location_id))  +
  labs(title = "PCA of filtered genetic variation - North American wild individuals",
       subtitle = "coloured by river (149 unique),  5455 individuals, 65576 SNPs")+
  xlab("PC1 (3.39% of variance)") +
  ylab("PC2 (1.46% variance explained)")+
  theme(legend.position = "none")

ggsave('../../../external/wild_NA_only_pca_by_location.png', location_pca_of_filtered_NA_wild)


Prov_pca_of_filtered_NA_wild = ggplot() + 
  geom_point(data = NA_PC_Meta, aes(x =PC1, y = PC2, colour = Province))  +
  labs(title = "PCA of filtered genetic variation - North American wild individuals",
       subtitle = "coloured by Province/State,  5455 individuals, 65576 SNPs")+
  xlab("PC1 (3.39% of variance)") +
  ylab("PC2 (1.46% variance explained)")

ggsave('../../../external/wild_NA_only_pca_by_Province.png', Prov_pca_of_filtered_NA_wild)


############################
# 3. run the admixture

print("ask tony what size of k to use?")
print("I ran 2-4, Tony had suggested 4")
## admixture wildNA_220k_filtered.bed 4



###########################
# 4. plot the admixture

print("tODO - read in the admixture values and make the plots for differening K's")
print("need decisison on how to panel things for just the NA wild (by province or something?/)")


#join this with the different Q sets
#NA_baseline_Meta

NA_baseline_Meta = read_tsv('filtered_wild_individuals_meta_data.tsv')
table(NA_baseline_Meta$Province)

NA_Q2_file = 'wildNA_220k_filtered.2.Q'
NA_Q3_file = 'wildNA_220k_filtered.3.Q'
NA_Q4_file = 'wildNA_220k_filtered.4.Q'


AdmixQ2_data = fread(NA_Q2_file)
colnames(AdmixQ2_data) = c("Q1", "Q2")
dim(AdmixQ2_data) #5455, proper number of individuals

AdmixQ3_data = fread(NA_Q3_file)
colnames(AdmixQ3_data) = c("Q1", "Q2", "Q3")
dim(AdmixQ3_data) #5455, proper number of individuals

AdmixQ4_data = fread(NA_Q4_file)
colnames(AdmixQ4_data) = c("Q1", "Q2", "Q3", "Q4")
dim(AdmixQ4_data) #5455, proper number of individuals


dim(NA_baseline_Meta) #5455 so matching indiviudal sets
table(NA_baseline_Meta$Province)


print("plot the different k by province or similar")

################################################################################
print("k=2")
meta_k2_Q_vals = cbind(NA_baseline_Meta, AdmixQ2_data)
meta_k2_Q_vals = meta_k2_Q_vals[meta_k2_Q_vals$Province != "NA", ]

k2_qlong = gather(meta_k2_Q_vals, Q_type, Q, Q1, Q2) 
k2_qlong =k2_qlong[order(k2_qlong$data_class, k2_qlong$location_id),]
k2_qlong$data_class = as.factor(k2_qlong$Province)

print("can change order by sitecode to something else, i.e. province")
k2_admix_bar = ggplot(data=k2_qlong, aes(x=reorder(fish_id, SiteCode), y=Q, fill=Q_type))+
  geom_bar(stat = "identity")+
  scale_y_continuous(expand = c(0,0))+
  ylab("Admixture proportion")+
  labs(title="Admixture of NA wild samples", 
       subtitle = "5455 NA fish only, K=2")+
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.x = element_blank(),
        strip.text = element_text(colour="black", size=12),
        panel.grid = element_blank(),
        panel.background = element_blank(),
        legend.title = element_blank(),
        legend.text = element_text(size = 12))


ggsave('../../../external/NA_wild_k2_ADMIX.png', k2_admix_bar)


print("faceted plot")
k2_panel_facet_admix_bar = ggplot(data=k2_qlong, aes(x=reorder(fish_id, SiteCode), y=Q, fill=Q_type))+
  geom_bar(stat = "identity")+
  scale_y_continuous(expand = c(0,0))+
  facet_wrap(~data_class, scales = "free", ncol = 2)+
  ylab("Admixture proportion K=2")+
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.x = element_blank(),
        strip.text = element_text(colour="black", size=12),
        panel.grid = element_blank(),
        panel.background = element_blank(),
        legend.title = element_blank(),
        legend.text = element_text(size = 12))

ggsave(paste0('../../../external/wild_na_faceted_province_ADMIX_k2.png'), k2_panel_facet_admix_bar)





################################################################################
print("k=3")
meta_k3_Q_vals = cbind(NA_baseline_Meta, AdmixQ3_data)

meta_k3_Q_vals = meta_k3_Q_vals[meta_k3_Q_vals$Province != "NA", ]

k3_qlong = gather(meta_k3_Q_vals, Q_type, Q, Q1, Q2, Q3) 
k3_qlong =k3_qlong[order(k3_qlong$data_class, k3_qlong$location_id),]
k3_qlong$data_class = as.factor(k3_qlong$Province)

print("can change order by sitecode to something else, i.e. province")
k3_admix_bar = ggplot(data=k3_qlong, aes(x=reorder(fish_id, data_class), y=Q, fill=Q_type))+
  geom_bar(stat = "identity")+
  scale_y_continuous(expand = c(0,0))+
  ylab("Admixture proportion")+
  labs(title="Admixture of NA wild samples", 
       subtitle = "5455 NA fish only, K=3")+
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.x = element_blank(),
        strip.text = element_text(colour="black", size=12),
        panel.grid = element_blank(),
        panel.background = element_blank(),
        legend.title = element_blank(),
        legend.text = element_text(size = 12))


ggsave('../../../external/NA_wild_k3_ADMIX.png', k2_admix_bar)


print("faceted plot")
k3_panel_facet_admix_bar = ggplot(data=k3_qlong, aes(x=reorder(fish_id, SiteCode), y=Q, fill=Q_type))+
  geom_bar(stat = "identity")+
  scale_y_continuous(expand = c(0,0))+
  facet_wrap(~data_class, scales = "free", ncol = 2)+
  ylab("Admixture proportion")+
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.x = element_blank(),
        strip.text = element_text(colour="black", size=12),
        panel.grid = element_blank(),
        panel.background = element_blank(),
        legend.title = element_blank(),
        legend.text = element_text(size = 12))

ggsave(paste0('../../../external/wild_na_faceted_province_ADMIX_k3.png'), k3_panel_facet_admix_bar)



################################################################################
print("k=4")
meta_k4_Q_vals = cbind(NA_baseline_Meta, AdmixQ4_data)


meta_k4_Q_vals = meta_k4_Q_vals[meta_k4_Q_vals$Province != "NA", ]

k4_qlong = gather(meta_k4_Q_vals, Q_type, Q, Q1, Q2, Q3, Q4) 
k4_qlong =k4_qlong[order(k4_qlong$data_class, k4_qlong$location_id),]
k4_qlong$data_class = as.factor(k4_qlong$Province)

print("can change order by sitecode to something else, i.e. province")
k4_admix_bar = ggplot(data=k4_qlong, aes(x=reorder(fish_id, data_class), y=Q, fill=Q_type))+
  geom_bar(stat = "identity")+
  scale_y_continuous(expand = c(0,0))+
  ylab("Admixture proportion")+
  labs(title="Admixture of NA wild samples", 
       subtitle = "5455 NA fish only, K=4")+
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.x = element_blank(),
        strip.text = element_text(colour="black", size=12),
        panel.grid = element_blank(),
        panel.background = element_blank(),
        legend.title = element_blank(),
        legend.text = element_text(size = 12))


ggsave('../../../external/NA_wild_k4_ADMIX.png', k4_admix_bar)


print("faceted plot")
k4_panel_facet_admix_bar = ggplot(data=k4_qlong, aes(x=reorder(fish_id, SiteCode), y=Q, fill=Q_type))+
  geom_bar(stat = "identity")+
  scale_y_continuous(expand = c(0,0))+
  facet_wrap(~data_class, scales = "free", ncol = 2)+
  ylab("Admixture proportion (k = 4)")+
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.x = element_blank(),
        strip.text = element_text(colour="black", size=12),
        panel.grid = element_blank(),
        panel.background = element_blank(),
        legend.title = element_blank(),
        legend.text = element_text(size = 12))

ggsave(paste0('../../../external/wild_na_faceted_province_ADMIX_k4.png'), k4_panel_facet_admix_bar)












