

library(tidyverse)

#' Read a ped and map file into R, produce a dataframe for general use
readPedMap_tsv_fmt = function(ped_file, map_file = "", headers = FALSE){
  if(headers == TRUE){
    #
    map_data = read_tsv(map_file, skip = 1, col_names = c("chromosome", "snp" , "genetic_distance", "physical_distance"))
    
    header_data = c('#family', 'individual', 'sire', 'dam', 'sex', 'pheno')
    header_data = c(header_data, map_data$snp)
    
    snp_data = read_tsv(ped_file, skip = 1, col_names = header_data)
    return(snp_data)
    
  }else{
    
    map_data = read_tsv(map_file, col_names = c("chromosome", "snp" , "genetic_distance", "physical_distance"))
    
    header_data = c('#family', 'individual', 'sire', 'dam', 'sex', 'pheno')
    header_data = c(header_data, map_data$snp)
    
    snp_data = read_tsv(ped_file, col_names = header_data)
    return(snp_data)
  }
  
}



###################
# run the analysis
library(tidyverse)
library(data.table)

setwd("/where/the/stuff/is/")
#for me
#setwd("C:/Users/camnu/bin/salmon-genomic-vulnerability/data/run_timing")

ped_file = 'maf_filtered_Salmo_220K_2022_RivAge_24.ped'
map_file = 'maf_filtered_Salmo_220K_2022_RivAge_24.map'

#note this is a really big one ~1.7Gb, you probably don't want the whole thing in R ever
snp_data = readPedMap_tsv_fmt(ped_file, map_file)
snp_data = as_tibble(snp_data)

snp_columns = names(snp_data)[7:length(names(snp_data))]
snp_columns