# TODO 

[x] - generate a list of the old files that need to be updated.
        - i.e. the LD inputs, Fst inputs, go through the scripts to see what was used

[x] - where needed, convert the BED to PED/MAP
    
[x] - update the map files

[x] - where needed, turn the new map and PED files into a new BED file

[] - redo the analyses
    [x] - fst
    [x] - LD
    [] - gene comparison



## Files to update

### Fst
#location 1:
full_fst_files = c("raw/fst_files/LAB_v_NFL_full_pairwise_fst_out.fst", 
                      "raw/fst_files/LAB_v_STL_full_pairwise_fst_out.fst", 
                      "raw/fst_files/NFL_v_STL_full_pairwise_fst_out.fst")
names(full_fst_files) = comps

matched_fst_files = c("raw/fst_files/LAB_v_NFL_matched_pairwise_fst_out.fst", 
                      "raw/fst_files/LAB_v_STL_matched_pairwise_fst_out.fst", 
                      "raw/fst_files/NFL_v_STL_matched_pairwise_fst_out.fst")

#location 2:
write_tsv(all_sig_fst, "interim/all_pw_fst_gte_15.tsv")
#^change locations in that file!

In script 9, read in the new map file after the lines above, then update the chr and POS columns based on them and remake the plots and summary stats. Thats the only change needed!


### LD
BED FILES TO UPDATE

wildNA_220k_filtered
LD_files/Labrador_220k_filtered
LD_files/Newfoundland_220k_filtered
LD_files/STLawrence_220k_filtered

BED 
    -> PED/MAP 
        -> Change the map file, match to PED 
            -> Back to BED
                -> REDO the analysis
                    -> regenerate the outputs


### Heterozygosity
        -rerun with the new LD_files
        - make an additional set of files, one per location and then get the per location hwe
        - this will serve as the response for the next analysis (track sample size with the heterozygosity)




### Map update code
                print("TODO - here need the maps updated and resorted")
                new_map_file = "raw/genome_update/Ssa220K_v3_Genome.map"
                new_map = read_tsv(new_map_file)

                #' Take the previous output and replace the values in the 
                #' CHR and POS columns to relate to the new genome.
                update_positions = function(df, new_map){
                
                new_chr = new_map$CHR
                new_pos = as.numeric(new_map$POS)
                
                names(new_chr) = as.list(new_map$SNP)
                names(new_pos) = as.list(new_map$SNP)
                
                df$CHR = unlist(lapply(df$SNP , function(x){new_chr[[x]]}))
                df$POS = unlist(lapply(df$SNP , function(x){new_pos[[x]]}))
                
                df = df[order(df$CHR, df$POS),]
                
                return(df)  
                
                }

